

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gnomad.utils.annotations &mdash; gnomad master documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.utils.constants" href="constants.html" />
    <link rel="prev" title="gnomad.utils" href="index.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> gnomad
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">utils</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="constants.html">constants</a></li>
<li class="toctree-l3"><a class="reference internal" href="generic.html">generic</a></li>
<li class="toctree-l3"><a class="reference internal" href="gnomad_functions.html">gnomad_functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="ld.html">ld</a></li>
<li class="toctree-l3"><a class="reference internal" href="liftover.html">liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="relatedness.html">relatedness</a></li>
<li class="toctree-l3"><a class="reference internal" href="rf.html">rf</a></li>
<li class="toctree-l3"><a class="reference internal" href="sample_qc.html">sample_qc</a></li>
<li class="toctree-l3"><a class="reference internal" href="slack.html">slack</a></li>
<li class="toctree-l3"><a class="reference internal" href="sparse_mt.html">sparse_mt</a></li>
<li class="toctree-l3"><a class="reference internal" href="variant_qc.html">variant_qc</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">gnomad</a> &raquo;</li>
        
          <li><a href="index.html">gnomad.utils</a> &raquo;</li>
        
      <li>gnomad.utils.annotations</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api_reference/utils/annotations.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gnomad-utils-annotations">
<h1>gnomad.utils.annotations<a class="headerlink" href="#gnomad-utils-annotations" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.pop_max_expr" title="gnomad.utils.annotations.pop_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.pop_max_expr</span></code></a>(freq, …)</p></td>
<td><p>Creates an expression containing popmax: the frequency information about the population that has the highest AF from the populations provided in <cite>freq_meta</cite>, excluding those specified in <cite>pops_to_exclude</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.project_max_expr" title="gnomad.utils.annotations.project_max_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.project_max_expr</span></code></a>(…)</p></td>
<td><p>Creates an expression that computes allele frequency information by project for the <cite>n_projects</cite> with the largest AF at this row.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.faf_expr" title="gnomad.utils.annotations.faf_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.faf_expr</span></code></a>(freq, …)</p></td>
<td><p>Calculates the filtering allele frequency (FAF) for each threshold specified in <cite>faf_thresholds</cite>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.qual_hist_expr" title="gnomad.utils.annotations.qual_hist_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.qual_hist_expr</span></code></a>([…])</p></td>
<td><p>Returns a struct expression with genotype quality histograms based on the arguments given (dp, gq, ad).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.age_hists_expr" title="gnomad.utils.annotations.age_hists_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.age_hists_expr</span></code></a>(…)</p></td>
<td><p>Returns a StructExpression with the age histograms for hets and homs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.annotate_freq" title="gnomad.utils.annotations.annotate_freq"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.annotate_freq</span></code></a>(mt[, …])</p></td>
<td><p>Adds a row annotation <cite>freq</cite> to the input <cite>mt</cite> with stratified allele frequencies, and a global annotation <cite>freq_meta</cite> with metadata.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.get_lowqual_expr" title="gnomad.utils.annotations.get_lowqual_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.get_lowqual_expr</span></code></a>(…)</p></td>
<td><p>Computes lowqual threshold expression for either split or unsplit alleles based on QUALapprox or AS_QUALapprox</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.generate_trio_stats_expr" title="gnomad.utils.annotations.generate_trio_stats_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.generate_trio_stats_expr</span></code></a>(trio_mt)</p></td>
<td><p>Generates a row-wise expression containing the following counts:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.filter_mt_to_trios" title="gnomad.utils.annotations.filter_mt_to_trios"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.filter_mt_to_trios</span></code></a>(mt, …)</p></td>
<td><p>Filters a MatrixTable to a set of trios in <cite>fam_ht</cite>, filters to autosomes, and annotates with adj.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.default_generate_trio_stats" title="gnomad.utils.annotations.default_generate_trio_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.default_generate_trio_stats</span></code></a>(mt)</p></td>
<td><p>Default function to run <cite>generate_trio_stats_expr</cite> to get trio stats stratified by raw and adj</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.generate_sib_stats_expr" title="gnomad.utils.annotations.generate_sib_stats_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.generate_sib_stats_expr</span></code></a>(mt, …)</p></td>
<td><p>Generates a row-wise expression containing the number of alternate alleles in common between sibling pairs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.default_generate_sib_stats" title="gnomad.utils.annotations.default_generate_sib_stats"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.default_generate_sib_stats</span></code></a>(mt, …)</p></td>
<td><p>This is meant as a default wrapper for <cite>generate_sib_stats_expr</cite>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.annotations.get_annotations_hists" title="gnomad.utils.annotations.get_annotations_hists"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.get_annotations_hists</span></code></a>(ht, …)</p></td>
<td><p>Creates histograms for variant metrics in ht.info.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.annotations.create_frequency_bins_expr" title="gnomad.utils.annotations.create_frequency_bins_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.annotations.create_frequency_bins_expr</span></code></a>(AC, AF)</p></td>
<td><p>Creates bins for frequencies in preparation for aggregating QUAL by frequency bin.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.utils.annotations"></span><dl class="function">
<dt id="gnomad.utils.annotations.pop_max_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">pop_max_expr</code><span class="sig-paren">(</span><em class="sig-param">freq</em>, <em class="sig-param">freq_meta</em>, <em class="sig-param">pops_to_exclude=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#pop_max_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.pop_max_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an expression containing popmax: the frequency information about the population
that has the highest AF from the populations provided in <cite>freq_meta</cite>,
excluding those specified in <cite>pops_to_exclude</cite>.
Only frequencies from adj populations are considered.</p>
<p>This resulting struct contains the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
<li><p>pop: str</p></li>
</ul>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>freq</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of Structs with fields [‘AC’, ‘AF’, ‘AN’, ‘homozygote_count’]</p></li>
<li><p><strong>freq_meta</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of meta dictionaries corresponding to freq (as returned by annotate_freq)</p></li>
<li><p><strong>pops_to_exclude</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Set of populations to skip for popmax calcluation</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Popmax struct</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.project_max_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">project_max_expr</code><span class="sig-paren">(</span><em class="sig-param">project_expr</em>, <em class="sig-param">gt_expr</em>, <em class="sig-param">alleles_expr</em>, <em class="sig-param">n_projects=5</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#project_max_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.project_max_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an expression that computes allele frequency information by project for the <cite>n_projects</cite> with the largest AF at this row.
This return an array with one element per non-reference allele.</p>
<p>Each of these elements is itself an array of structs with the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
<li><p>project: str</p></li>
</ul>
</div></blockquote>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Only projects with AF &gt; 0 are returned.
In case of ties, the project ordering is not guaranteed, and at most <cite>n_projects</cite> are returned.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>project_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>) – column expression containing the project</p></li>
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – entry expression containing the genotype</p></li>
<li><p><strong>alleles_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – row expression containing the alleles</p></li>
<li><p><strong>n_projects</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Maximum number of projects to return for each row</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>projectmax expression</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.faf_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">faf_expr</code><span class="sig-paren">(</span><em class="sig-param">freq, freq_meta, locus, pops_to_exclude=None, faf_thresholds=[0.95, 0.99]</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#faf_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.faf_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the filtering allele frequency (FAF) for each threshold specified in <cite>faf_thresholds</cite>.
See <a class="reference external" href="http://cardiodb.org/allelefrequencyapp/">http://cardiodb.org/allelefrequencyapp/</a> for more information.</p>
<p>The FAF is computed for each of the following population stratification if found in <cite>freq_meta</cite>:</p>
<blockquote>
<div><ul class="simple">
<li><p>All samples, with adj criteria</p></li>
<li><p>For each population, with adj criteria</p></li>
<li><p>For all sex/population on the non-PAR regions of sex chromosomes (will be missing on autosomes and PAR regions of sex chromosomes)</p></li>
</ul>
</div></blockquote>
<p>Each of the FAF entry is a struct with one entry per threshold specified in <cite>faf_thresholds</cite> of type float64.</p>
<p>This returns a tuple with two expressions:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>An array of FAF expressions as described above</p></li>
<li><p>An array of dict containing the metadata for each of the array elements, in the same format as that produced by <cite>annotate_freq</cite>.</p></li>
</ol>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>freq</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of call stats structs (typically generated by hl.agg.call_stats)</p></li>
<li><p><strong>freq_meta</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – ArrayExpression of meta dictionaries corresponding to freq (typically generated using annotate_freq)</p></li>
<li><p><strong>locus</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.LocusExpression.html#hail.expr.LocusExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">LocusExpression</span></code></a>) – locus</p></li>
<li><p><strong>pops_to_exclude</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – Set of populations to exclude from faf calculation (typically bottlenecked or consanguineous populations)</p></li>
<li><p><strong>faf_thresholds</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – List of FAF thresholds to compute</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>(FAF expression, FAF metadata)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.qual_hist_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">qual_hist_expr</code><span class="sig-paren">(</span><em class="sig-param">gt_expr=None</em>, <em class="sig-param">gq_expr=None</em>, <em class="sig-param">dp_expr=None</em>, <em class="sig-param">ad_expr=None</em>, <em class="sig-param">adj_expr=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#qual_hist_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.qual_hist_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a struct expression with genotype quality histograms based on the arguments given (dp, gq, ad).</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>If <cite>gt_expr</cite> is provided, will return histograms for non-reference samples only as well as all samples.</p></li>
<li><p><cite>gt_expr</cite> is required for the allele-balance histogram, as it is only computed on het samples.</p></li>
<li><p>If <cite>adj_expr</cite> is provided, additional histograms are computed using only adj samples.</p></li>
</ul>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gt_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>]) – Entry expression containing genotype</p></li>
<li><p><strong>gq_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – Entry expression containing genotype quality</p></li>
<li><p><strong>dp_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – Entry expression containing depth</p></li>
<li><p><strong>ad_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>]) – Entry expression containing allelic depth (bi-allelic here)</p></li>
<li><p><strong>adj_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Entry expression containing adj (high quality) genotype status</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Genotype quality histograms expression</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.age_hists_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">age_hists_expr</code><span class="sig-paren">(</span><em class="sig-param">adj_expr</em>, <em class="sig-param">gt_expr</em>, <em class="sig-param">age_expr</em>, <em class="sig-param">lowest_boundary=30</em>, <em class="sig-param">highest_boundary=80</em>, <em class="sig-param">n_bins=10</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#age_hists_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.age_hists_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a StructExpression with the age histograms for hets and homs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>adj_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>) – Entry expression containing whether a genotype is high quality (adj) or not</p></li>
<li><p><strong>gt_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>) – Entry expression containing the genotype</p></li>
<li><p><strong>age_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Col expression containing the sample’s age</p></li>
<li><p><strong>lowest_boundary</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Lowest bin boundary (any younger sample will be binned in n_smaller)</p></li>
<li><p><strong>highest_boundary</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Highest bin boundary (any older sample will be binned in n_larger)</p></li>
<li><p><strong>n_bins</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Total number of bins</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A struct with <cite>age_hist_het</cite> and <cite>age_hist_hom</cite></p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.annotate_freq">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">annotate_freq</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">sex_expr=None</em>, <em class="sig-param">pop_expr=None</em>, <em class="sig-param">subpop_expr=None</em>, <em class="sig-param">additional_strata_expr=None</em>, <em class="sig-param">downsamplings=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#annotate_freq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.annotate_freq" title="Permalink to this definition">¶</a></dt>
<dd><p>Adds a row annotation <cite>freq</cite> to the input <cite>mt</cite> with stratified allele frequencies,
and a global annotation <cite>freq_meta</cite> with metadata.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently this only supports bi-allelic sites.
The input <cite>mt</cite> needs to have the following entry fields:
- GT: a CallExpression containing the genotype
- adj: a BooleanExpression containing whether the genotype is of high quality or not.
All expressions arguments need to be expression on the input <cite>mt</cite>.</p>
</div>
<p class="rubric"><cite>freq</cite> row annotation</p>
<p>The <cite>freq</cite> row annotation is an Array of Struct, with each Struct containing the following fields:</p>
<blockquote>
<div><ul class="simple">
<li><p>AC: int32</p></li>
<li><p>AF: float64</p></li>
<li><p>AN: int32</p></li>
<li><p>homozygote_count: int32</p></li>
</ul>
</div></blockquote>
<p>Each element of the array corresponds to a stratification of the data, and the metadata about these annotations is
stored in the globals.</p>
<p class="rubric">Global <cite>freq_meta</cite> metadata annotation</p>
<p>The global annotation <cite>freq_meta</cite> is added to the input <cite>mt</cite>. It is a list of dict.
Each element of the list contains metadata on a frequency stratification and the index in the list corresponds
to the index of that frequency stratification in the <cite>freq</cite> row annotation.</p>
<p class="rubric">The <cite>downsamplings</cite> parameter</p>
<p>If the <cite>downsamplings</cite> parameter is used, frequencies will be computed for all samples and by population
(if <cite>pop_expr</cite> is specified) by downsampling the number of samples without replacement to each of the numbers specified in the
<cite>downsamplings</cite> array, provided that there are enough samples in the dataset.
In addition, if <cite>pop_expr</cite> is specified, a downsampling to each of the exact number of samples present in each population is added.
Note that samples are randomly sampled only once, meaning that the lower downsamplings are subsets of the higher ones.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable</p></li>
<li><p><strong>sex_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by sex. If <cite>pop_expr</cite> is also specified, then a pop/sex stratifiction is added.</p></li>
<li><p><strong>pop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by population. If <cite>sex_expr</cite> is also specified, then a pop/sex stratifiction is added.</p></li>
<li><p><strong>subpop_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]) – When specified, frequencies are stratified by sub-continental population. Note that <cite>pop_expr</cite> is required as well when using this option.</p></li>
<li><p><strong>additional_strata_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StringExpression.html#hail.expr.StringExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StringExpression</span></code></a>]]) – When specified, frequencies are stratified by the given additional strata found in the dict. This can e.g. be used to stratify by platform.</p></li>
<li><p><strong>downsamplings</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]) – When specified, frequencies are computed by downsampling the data to the number of samples given in the list. Note that if <cite>pop_expr</cite> is specified, downsamplings by population is also computed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>MatrixTable with <cite>freq</cite> annotation</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.get_lowqual_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">get_lowqual_expr</code><span class="sig-paren">(</span><em class="sig-param">alleles</em>, <em class="sig-param">qual_approx_expr</em>, <em class="sig-param">snv_phred_threshold=30</em>, <em class="sig-param">snv_phred_het_prior=30</em>, <em class="sig-param">indel_phred_threshold=30</em>, <em class="sig-param">indel_phred_het_prior=39</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#get_lowqual_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.get_lowqual_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes lowqual threshold expression for either split or unsplit alleles based on QUALapprox or AS_QUALapprox</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This lowqual annotation differs from the GATK LowQual filter. GATK uses the SNV prior if the site has any SNV
so that SNVs are not penalized if there are also indels at those sites. This implementation will use the indel
threshold for indels and the SNV threshold for the SNVs therefore excluding more indels than the GATK filter.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alleles</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>) – Array of alleles</p></li>
<li><p><strong>qual_approx_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – QUALapprox or AS_QUALapprox</p></li>
<li><p><strong>snv_phred_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled SNV “emission” threshold (similar to GATK emission threshold)</p></li>
<li><p><strong>snv_phred_het_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled SNV heterozygosity prior (30 = 1/1000 bases, GATK default)</p></li>
<li><p><strong>indel_phred_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled indel “emission” threshold (similar to GATK emission threshold)</p></li>
<li><p><strong>indel_phred_het_prior</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Phred-scaled indel heterozygosity prior (30 = 1/1000 bases, GATK default)</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayExpression.html#hail.expr.ArrayExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayExpression</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>lowqual expression (BooleanExpression if <cite>qual_approx_expr`is Numeric, Array[BooleanExpression] if `qual_approx_expr</cite> is ArrayNumeric)</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.generate_trio_stats_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">generate_trio_stats_expr</code><span class="sig-paren">(</span><em class="sig-param">trio_mt</em>, <em class="sig-param">transmitted_strata={'raw': True}</em>, <em class="sig-param">de_novo_strata={'raw': True}</em>, <em class="sig-param">ac_strata={'raw': True}</em>, <em class="sig-param">proband_is_female_expr=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#generate_trio_stats_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.generate_trio_stats_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a row-wise expression containing the following counts:</p>
<blockquote>
<div><ul class="simple">
<li><p>Number of alleles in het parents transmitted to the proband</p></li>
<li><p>Number of alleles in het parents not transmitted to the proband</p></li>
<li><p>Number of de novo mutations</p></li>
<li><p>Parent allele count</p></li>
<li><p>Proband allele count</p></li>
</ul>
</div></blockquote>
<p>Transmission and de novo mutation metrics and allele counts can be stratified using additional filters.
<cite>transmitted_strata</cite>, <cite>de_novo_strata</cite>, and <cite>ac_strata</cite> all expect a dictionary of filtering expressions keyed
by their desired suffix to append for labeling. The default will perform counts using all genotypes and append
‘raw’ to the label.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Expects that <cite>mt</cite> is dense if dealing with a sparse MT <cite>hl.experimental.densify</cite> must be run first.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>trio_mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – A trio standard trio MT (with the format as produced by hail.methods.trio_matrix)</p></li>
<li><p><strong>transmitted_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the transmission counts</p></li>
<li><p><strong>de_novo_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the de novo counts</p></li>
<li><p><strong>ac_strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata for the parent and child allele counts</p></li>
<li><p><strong>proband_is_female_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – An optional expression giving the sex the proband. If not given, DNMs are only computed for autosomes.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An expression with the counts</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.filter_mt_to_trios">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">filter_mt_to_trios</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">fam_ht</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#filter_mt_to_trios"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.filter_mt_to_trios" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters a MatrixTable to a set of trios in <cite>fam_ht</cite>, filters to autosomes, and annotates with adj.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – A Matrix Table to filter to only trios</p></li>
<li><p><strong>fam_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – A Table of trios to filter to, loaded using <cite>hl.import_fam</cite></p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A MT filtered to trios and adj annotated</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.default_generate_trio_stats">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">default_generate_trio_stats</code><span class="sig-paren">(</span><em class="sig-param">mt</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#default_generate_trio_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.default_generate_trio_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>Default function to run <cite>generate_trio_stats_expr</cite> to get trio stats stratified by raw and adj</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Expects that <cite>mt</cite> is it a trio matrix table that was annotated with adj and if dealing with
a sparse MT <cite>hl.experimental.densify</cite> must be run first.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – A Trio Matrix Table returned from <cite>hl.trio_matrix</cite>. Must be dense</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Table with trio stats</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.generate_sib_stats_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">generate_sib_stats_expr</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">sib_ht</em>, <em class="sig-param">i_col='i'</em>, <em class="sig-param">j_col='j'</em>, <em class="sig-param">strata={'raw': True}</em>, <em class="sig-param">is_female=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#generate_sib_stats_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.generate_sib_stats_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a row-wise expression containing the number of alternate alleles in common between sibling pairs.</p>
<p>The sibling sharing counts can be stratified using additional filters using <cite>stata</cite>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function expects that the <cite>mt</cite> has either been split or filtered to only bi-allelics
If a sample has multiple sibling pairs, only one pair will be counted</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input matrix table</p></li>
<li><p><strong>sib_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table defining sibling pairs with one sample in a col (<cite>i_col</cite>) and the second in another col (<cite>j_col</cite>)</p></li>
<li><p><strong>i_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 1st sample of the pair in the relationship table</p></li>
<li><p><strong>j_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 2nd sample of the pair in the relationship table</p></li>
<li><p><strong>strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Dict with additional strata to use when computing shared sibling variant counts</p></li>
<li><p><strong>is_female</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – An optional column in mt giving the sample sex. If not given, counts are only computed for autosomes.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with the sibling shared variant counts</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.default_generate_sib_stats">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">default_generate_sib_stats</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">relatedness_ht</em>, <em class="sig-param">sex_ht</em>, <em class="sig-param">i_col='i'</em>, <em class="sig-param">j_col='j'</em>, <em class="sig-param">relationship_col='relationship'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#default_generate_sib_stats"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.default_generate_sib_stats" title="Permalink to this definition">¶</a></dt>
<dd><p>This is meant as a default wrapper for <cite>generate_sib_stats_expr</cite>. It returns a hail table with counts of variants
shared by pairs of siblings in <cite>relatedness_ht</cite>.</p>
<p>This function takes a hail Table with a row for each pair of individuals i,j in the data that are related (it’s OK to have unrelated samples too).
The <cite>relationship_col</cite> should be a column specifying the relationship between each two samples as defined by
the constants in <cite>gnomad.utils.relatedness</cite>. This relationship_col will be used to filter to only pairs of
samples that are annotated as <cite>SIBLINGS</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input Matrix table</p></li>
<li><p><strong>relatedness_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input relationship table</p></li>
<li><p><strong>sex_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – A Table containing sex information for the samples</p></li>
<li><p><strong>i_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 1st sample of the pair in the relationship table</p></li>
<li><p><strong>j_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the 2nd sample of the pair in the relationship table</p></li>
<li><p><strong>relationship_col</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Column containing the relationship for the sample pair as defined in this module constants.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with the sibling shared variant counts</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.get_annotations_hists">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">get_annotations_hists</code><span class="sig-paren">(</span><em class="sig-param">ht, annotations_hists, log10_annotations=['DP']</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#get_annotations_hists"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.get_annotations_hists" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates histograms for variant metrics in ht.info.
Used when creating site quality distribution json files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with variant metrics</p></li>
<li><p><strong>annotations_hists</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>]) – Dictionary of metrics names and their histogram values (start, end, bins)</p></li>
<li><p><strong>log10_annotations</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – List of metrics to log scale</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Dictionary of merics and their histograms</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Dict[str, hl.expr.StructExpression]</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.annotations.create_frequency_bins_expr">
<code class="sig-prename descclassname">gnomad.utils.annotations.</code><code class="sig-name descname">create_frequency_bins_expr</code><span class="sig-paren">(</span><em class="sig-param">AC</em>, <em class="sig-param">AF</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/annotations.html#create_frequency_bins_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.annotations.create_frequency_bins_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates bins for frequencies in preparation for aggregating QUAL by frequency bin.</p>
<dl class="simple">
<dt>Bins: </dt><dd><ul class="simple">
<li><p>singleton</p></li>
<li><p>doubleton</p></li>
<li><p>0.00005</p></li>
<li><p>0.0001</p></li>
<li><p>0.0002</p></li>
<li><p>0.0005</p></li>
<li><p>0.001,</p></li>
<li><p>0.002</p></li>
<li><p>0.005</p></li>
<li><p>0.01</p></li>
<li><p>0.02</p></li>
<li><p>0.05</p></li>
<li><p>0.1</p></li>
<li><p>0.2</p></li>
<li><p>0.5</p></li>
<li><p>1</p></li>
</ul>
</dd>
</dl>
<p>NOTE: Frequencies should be frequencies from raw data.
Used when creating site quality distribution json files.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>AC</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Field in input that contains the allele count information</p></li>
<li><p><strong>AF</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Field in input that contains the allele frequency information</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Expression containing bin name</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>hl.expr.StringExpression</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="constants.html" class="btn btn-neutral float-right" title="gnomad.utils.constants" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral float-left" title="gnomad.utils" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>