

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gnomad.utils.sample_qc &mdash; gnomad master documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="gnomad.utils.slack" href="slack.html" />
    <link rel="prev" title="gnomad.utils.rf" href="rf.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> gnomad
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">API Reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../resources/index.html">resources</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">utils</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="annotations.html">annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="constants.html">constants</a></li>
<li class="toctree-l3"><a class="reference internal" href="generic.html">generic</a></li>
<li class="toctree-l3"><a class="reference internal" href="gnomad_functions.html">gnomad_functions</a></li>
<li class="toctree-l3"><a class="reference internal" href="ld.html">ld</a></li>
<li class="toctree-l3"><a class="reference internal" href="liftover.html">liftover</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="relatedness.html">relatedness</a></li>
<li class="toctree-l3"><a class="reference internal" href="rf.html">rf</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">sample_qc</a></li>
<li class="toctree-l3"><a class="reference internal" href="slack.html">slack</a></li>
<li class="toctree-l3"><a class="reference internal" href="sparse_mt.html">sparse_mt</a></li>
<li class="toctree-l3"><a class="reference internal" href="variant_qc.html">variant_qc</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">gnomad</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../index.html">gnomad</a> &raquo;</li>
        
          <li><a href="index.html">gnomad.utils</a> &raquo;</li>
        
      <li>gnomad.utils.sample_qc</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api_reference/utils/sample_qc.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="gnomad-utils-sample-qc">
<h1>gnomad.utils.sample_qc<a class="headerlink" href="#gnomad-utils-sample-qc" title="Permalink to this headline">¶</a></h1>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.filter_rows_for_qc" title="gnomad.utils.sample_qc.filter_rows_for_qc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.filter_rows_for_qc</span></code></a>(mt)</p></td>
<td><p>Annotates rows with <cite>sites_callrate</cite>, <cite>site_inbreeding_coeff</cite> and <cite>af</cite>, then applies thresholds.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.get_qc_mt" title="gnomad.utils.sample_qc.get_qc_mt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.get_qc_mt</span></code></a>(mt[, …])</p></td>
<td><p>Creates a QC-ready MT by keeping:</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.compute_callrate_mt" title="gnomad.utils.sample_qc.compute_callrate_mt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.compute_callrate_mt</span></code></a>(mt, …)</p></td>
<td><p>Computes a sample/interval MT with each entry containing the call rate for that sample/interval.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.run_platform_pca" title="gnomad.utils.sample_qc.run_platform_pca"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.run_platform_pca</span></code></a>(…)</p></td>
<td><p>Runs a PCA on a sample/interval MT with each entry containing the call rate.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.assign_platform_from_pcs" title="gnomad.utils.sample_qc.assign_platform_from_pcs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.assign_platform_from_pcs</span></code></a>(…)</p></td>
<td><p>Assigns platforms using HBDSCAN on the results of call rate PCA.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.default_annotate_sex" title="gnomad.utils.sample_qc.default_annotate_sex"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.default_annotate_sex</span></code></a>(mt)</p></td>
<td><p>Imputes sample sex based on X-chromosome heterozygosity and sex chromosome ploidy.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.get_ploidy_cutoffs" title="gnomad.utils.sample_qc.get_ploidy_cutoffs"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.get_ploidy_cutoffs</span></code></a>(ht, …)</p></td>
<td><p>Gets chromosome X and Y ploidy cutoffs for XY and XX samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.get_sex_expr" title="gnomad.utils.sample_qc.get_sex_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.get_sex_expr</span></code></a>(…)</p></td>
<td><p>Creates a struct with the following annotation: - X_karyotype (str) - Y_karyotype (str) - sex_karyotype (str)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.compute_related_samples_to_drop" title="gnomad.utils.sample_qc.compute_related_samples_to_drop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.compute_related_samples_to_drop</span></code></a>(…)</p></td>
<td><p>Computes a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.run_pca_with_relateds" title="gnomad.utils.sample_qc.run_pca_with_relateds"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.run_pca_with_relateds</span></code></a>(…)</p></td>
<td><p>First runs PCA excluding the given related samples, then projects these samples in the PC space to return scores for all samples.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.compute_stratified_metrics_filter" title="gnomad.utils.sample_qc.compute_stratified_metrics_filter"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.compute_stratified_metrics_filter</span></code></a>(ht, …)</p></td>
<td><p>Compute median, MAD, and upper and lower thresholds for each metric used in outlier filtering</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.compute_qc_metrics_residuals" title="gnomad.utils.sample_qc.compute_qc_metrics_residuals"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.compute_qc_metrics_residuals</span></code></a>(ht, …)</p></td>
<td><p>Computes QC metrics residuals after regressing out PCs (and optionally PC^2)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.add_filters_expr" title="gnomad.utils.sample_qc.add_filters_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.add_filters_expr</span></code></a>(filters)</p></td>
<td><p>Creates an expression to create or add filters.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.merge_sample_qc_expr" title="gnomad.utils.sample_qc.merge_sample_qc_expr"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.merge_sample_qc_expr</span></code></a>(…)</p></td>
<td><p>Creates an expression that merges results from non-overlapping strata of hail.sample_qc</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#gnomad.utils.sample_qc.compute_stratified_sample_qc" title="gnomad.utils.sample_qc.compute_stratified_sample_qc"><code class="xref py py-obj docutils literal notranslate"><span class="pre">gnomad.utils.sample_qc.compute_stratified_sample_qc</span></code></a>(mt, …)</p></td>
<td><p>Runs hl.sample_qc on different strata and then also merge the results into a single expression.</p></td>
</tr>
</tbody>
</table>
<span class="target" id="module-gnomad.utils.sample_qc"></span><dl class="function">
<dt id="gnomad.utils.sample_qc.filter_rows_for_qc">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">filter_rows_for_qc</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">min_af=0.001</em>, <em class="sig-param">min_callrate=0.99</em>, <em class="sig-param">min_inbreeding_coeff_threshold=-0.8</em>, <em class="sig-param">min_hardy_weinberg_threshold=1e-08</em>, <em class="sig-param">apply_hard_filters=True</em>, <em class="sig-param">bi_allelic_only=True</em>, <em class="sig-param">snv_only=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#filter_rows_for_qc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.filter_rows_for_qc" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotates rows with <cite>sites_callrate</cite>, <cite>site_inbreeding_coeff</cite> and <cite>af</cite>, then applies thresholds.
AF and callrate thresholds are taken from gnomAD QC; inbreeding coeff, MQ, FS and QD filters are taken from GATK best practices</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function expect the typical <code class="docutils literal notranslate"><span class="pre">info</span></code> annotation of type struct with fields <code class="docutils literal notranslate"><span class="pre">MQ</span></code>, <code class="docutils literal notranslate"><span class="pre">FS</span></code> and <code class="docutils literal notranslate"><span class="pre">QD</span></code>
if applying hard filters.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MT</p></li>
<li><p><strong>min_af</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum site AF to keep. Not applied if set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>min_callrate</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum site call rate to keep. Not applied if set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>min_inbreeding_coeff_threshold</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum site inbreeding coefficient to keep. Not applied if set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>min_hardy_weinberg_threshold</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum site HW test p-value to keep. Not applied if set to <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p></li>
<li><p><strong>apply_hard_filters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
<li><p><strong>bi_allelic_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
<li><p><strong>snv_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Paramapply_hard_filters</dt>
<dd class="field-even"><p>Whether to apply standard GAKT default site hard filters: QD &gt;= 2, FS &lt;= 60 and MQ &gt;= 30</p>
</dd>
<dt class="field-odd">Parambi_allelic_only</dt>
<dd class="field-odd"><p>Whether to only keep bi-allelic sites or include multi-allelic sites too</p>
</dd>
<dt class="field-even">Paramsnv_only</dt>
<dd class="field-even"><p>Whether to only keep SNVs or include other variant types</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>annotated and filtered table</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.get_qc_mt">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">get_qc_mt</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">adj_only=True</em>, <em class="sig-param">min_af=0.001</em>, <em class="sig-param">min_callrate=0.99</em>, <em class="sig-param">min_inbreeding_coeff_threshold=-0.8</em>, <em class="sig-param">min_hardy_weinberg_threshold=1e-08</em>, <em class="sig-param">apply_hard_filters=True</em>, <em class="sig-param">ld_r2=0.1</em>, <em class="sig-param">filter_lcr=True</em>, <em class="sig-param">filter_decoy=True</em>, <em class="sig-param">filter_segdup=True</em>, <em class="sig-param">filter_exome_low_coverage_regions=False</em>, <em class="sig-param">high_conf_regions=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#get_qc_mt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.get_qc_mt" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a QC-ready MT by keeping:</p>
<ul class="simple">
<li><p>Variants outside known problematic regions</p></li>
<li><p>Bi-allelic SNVs only</p></li>
<li><p>Variants passing hard thresholds</p></li>
<li><p>Variants passing the set call rate and MAF thresholds</p></li>
<li><p>Genotypes passing on gnomAD ADJ criteria (GQ&gt;=20, DP&gt;=10, AB&gt;0.2 for hets)</p></li>
</ul>
<p>In addition, the MT will be LD-pruned if <cite>ld_r2</cite> is set.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MT</p></li>
<li><p><strong>adj_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, only ADJ genotypes are kept. This filter is applied before the call rate and AF calculation.</p></li>
<li><p><strong>min_af</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum allele frequency to keep</p></li>
<li><p><strong>min_callrate</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum call rate to keep</p></li>
<li><p><strong>min_inbreeding_coeff_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum site inbreeding coefficient to keep</p></li>
<li><p><strong>min_hardy_weinberg_threshold</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum site HW test p-value to keep</p></li>
<li><p><strong>apply_hard_filters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to apply standard GAKT default site hard filters: QD &gt;= 2, FS &lt;= 60 and MQ &gt;= 30</p></li>
<li><p><strong>ld_r2</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Minimum r2 to keep when LD-pruning (set to <cite>None</cite> for no LD pruning)</p></li>
<li><p><strong>filter_lcr</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Filter LCR regions</p></li>
<li><p><strong>filter_decoy</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Filter decoy regions</p></li>
<li><p><strong>filter_segdup</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Filter segmental duplication regions</p></li>
<li><p><strong>filter_exome_low_coverage_regions</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, only high coverage exome regions (computed from gnomAD are kept)</p></li>
<li><p><strong>high_conf_regions</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]) – If given, the data will be filtered to only include variants in those regions</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Filtered MT</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.compute_callrate_mt">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">compute_callrate_mt</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">intervals_ht</em>, <em class="sig-param">bi_allelic_only=True</em>, <em class="sig-param">autosomes_only=True</em>, <em class="sig-param">match=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#compute_callrate_mt"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.compute_callrate_mt" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a sample/interval MT with each entry containing the call rate for that sample/interval.
This can be used as input for imputing exome sequencing platforms.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The input interval HT should have a key of type Interval.
The resulting table will have a key of the same type as the <cite>intervals_ht</cite> table and
contain an <cite>interval_info</cite> field containing all non-key fields of the <cite>intervals_ht</cite>.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MT</p></li>
<li><p><strong>intervals_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table containing the intervals. This table has to be keyed by locus.</p></li>
<li><p><strong>bi_allelic_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, only bi-allelic sites are used for the computation</p></li>
<li><p><strong>autosomes_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – If set, only autosomal intervals are used.</p></li>
<li><p><strong>matches</strong> – If set, returns all intervals in intervals_ht that overlap the locus in the input MT.</p></li>
<li><p><strong>match</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Callrate MT</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.run_platform_pca">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">run_platform_pca</code><span class="sig-paren">(</span><em class="sig-param">callrate_mt</em>, <em class="sig-param">binarization_threshold=0.25</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#run_platform_pca"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.run_platform_pca" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs a PCA on a sample/interval MT with each entry containing the call rate.
When <cite>binzarization_threshold</cite> is set, the callrate is transformed to a 0/1 value based on the threshold.
E.g. with the default threshold of 0.25, all entries with a callrate &lt; 0.25 are considered as 0s, others as 1s.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>callrate_mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input callrate MT</p></li>
<li><p><strong>binarization_threshold</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – binzarization_threshold. None is no threshold desired</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>eigenvalues, scores_ht, loadings_ht</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.assign_platform_from_pcs">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">assign_platform_from_pcs</code><span class="sig-paren">(</span><em class="sig-param">platform_pca_scores_ht</em>, <em class="sig-param">pc_scores_ann='scores'</em>, <em class="sig-param">hdbscan_min_cluster_size=None</em>, <em class="sig-param">hdbscan_min_samples=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#assign_platform_from_pcs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.assign_platform_from_pcs" title="Permalink to this definition">¶</a></dt>
<dd><p>Assigns platforms using HBDSCAN on the results of call rate PCA.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>platform_pca_scores_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input table with the PCA score for each sample</p></li>
<li><p><strong>pc_scores_ann</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Field containing the scores</p></li>
<li><p><strong>hdbscan_min_cluster_size</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – HDBSCAN <cite>min_cluster_size</cite> parameter. If not specified the smallest of 500 and 0.1*n_samples will be used.</p></li>
<li><p><strong>hdbscan_min_samples</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – HDBSCAN <cite>min_samples</cite> parameter</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with a <cite>qc_platform</cite> annotation containing the platform based on HDBSCAN clustering</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.default_annotate_sex">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">default_annotate_sex</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">is_sparse=True</em>, <em class="sig-param">excluded_intervals=None</em>, <em class="sig-param">included_intervals=None</em>, <em class="sig-param">normalization_contig='chr20'</em>, <em class="sig-param">sites_ht=None</em>, <em class="sig-param">aaf_expr=None</em>, <em class="sig-param">gt_expr='GT'</em>, <em class="sig-param">f_stat_cutoff=0.5</em>, <em class="sig-param">aaf_threshold=0.001</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#default_annotate_sex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.default_annotate_sex" title="Permalink to this definition">¶</a></dt>
<dd><p>Imputes sample sex based on X-chromosome heterozygosity and sex chromosome ploidy.</p>
<dl class="simple">
<dt>Returns Table with the following fields:</dt><dd><ul class="simple">
<li><p>s (str): Sample</p></li>
<li><p>chr20_mean_dp (float32): Sample’s mean coverage over chromosome 20.</p></li>
<li><p>chrX_mean_dp (float32): Sample’s mean coverage over chromosome X.</p></li>
<li><p>chrY_mean_dp (float32): Sample’s mean coverage over chromosome Y.</p></li>
<li><p>chrX_ploidy (float32): Sample’s imputed ploidy over chromosome X.</p></li>
<li><p>chrY_ploidy (float32): Sample’s imputed ploidy over chromosome Y.</p></li>
<li><p>f_stat (float64): Sample f-stat. Calculated using hl.impute_sex.</p></li>
<li><p>n_called (int64): Number of variants with a genotype call. Calculated using hl.impute_sex.</p></li>
<li><p>expected_homs (float64): Expected number of homozygotes. Calculated using hl.impute_sex.</p></li>
<li><p>observed_homs (int64): Expected number of homozygotes. Calculated using hl.impute_sex.</p></li>
<li><p>X_karyotype (str): Sample’s chromosome X karyotype.</p></li>
<li><p>Y_karyotype (str): Sample’s chromosome Y karyotype.</p></li>
<li><p>sex_karyotype (str): Sample’s sex karyotype.</p></li>
</ul>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MatrixTable</p></li>
<li><p><strong>is_sparse</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether input MatrixTable is in sparse data format</p></li>
<li><p><strong>excluded_intervals</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Optional table of intervals to exclude from the computation.</p></li>
<li><p><strong>included_intervals</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Optional table of intervals to use in the computation. REQUIRED for exomes.</p></li>
<li><p><strong>normalization_contig</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Which chromosome to use to normalize sex chromosome coverage. Used in determining sex chromosome ploidies.</p></li>
<li><p><strong>sites_ht</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Optional Table to use. If present, filters input MatrixTable to sites in this Table prior to imputing sex,
and pulls alternate allele frequency from this Table.</p></li>
<li><p><strong>aaf_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional. Name of field in input MatrixTable with alternate allele frequency.</p></li>
<li><p><strong>gt_expr</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of entry field storing the genotype. Default: ‘GT’</p></li>
<li><p><strong>f_stat_cutoff</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – f-stat to roughly divide ‘XX’ from ‘XY’ samples. Assumes XX samples are below cutoff and XY are above cutoff.</p></li>
<li><p><strong>aaf_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Minimum alternate allele frequency to be used in f-stat calculations.</p></li>
<li><p><strong>is_sparse</strong> – </p></li>
<li><p><strong>aaf_threshold</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Table of samples and their imputed sex karyotypes.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.get_ploidy_cutoffs">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">get_ploidy_cutoffs</code><span class="sig-paren">(</span><em class="sig-param">ht</em>, <em class="sig-param">f_stat_cutoff</em>, <em class="sig-param">normal_ploidy_cutoff=5</em>, <em class="sig-param">aneuploidy_cutoff=6</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#get_ploidy_cutoffs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.get_ploidy_cutoffs" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets chromosome X and Y ploidy cutoffs for XY and XX samples. Note this assumes the input hail Table has the fields f_stat, chrX_ploidy, and chrY_ploidy.
Returns a tuple of sex chromosome ploidy cutoffs: ((x_ploidy_cutoffs), (y_ploidy_cutoffs)).
x_ploidy_cutoffs: (upper cutoff for single X, (lower cutoff for double X, upper cutoff for double X), lower cutoff for triple X)
y_ploidy_cutoffs: ((lower cutoff for single Y, upper cutoff for single Y), lower cutoff for double Y)</p>
<p>Uses the normal_ploidy_cutoff parameter to determine the ploidy cutoffs for XX and XY karyotypes.
Uses the aneuploidy_cutoff parameter to determine the cutoffs for sex aneuploidies.</p>
<p>Note that f-stat is used only to split the samples into roughly ‘XX’ and ‘XY’ categories and is not used in the final karyotype annotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with f_stat and sex chromosome ploidies</p></li>
<li><p><strong>f_stat_cutoff</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – f-stat to roughly divide ‘XX’ from ‘XY’ samples. Assumes XX samples are below cutoff and XY are above cutoff.</p></li>
<li><p><strong>normal_ploidy_cutoff</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of standard deviations to use when determining sex chromosome ploidy cutoffs for XX, XY karyotypes.</p></li>
<li><p><strong>aneuploidy_cutoff</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of standard deviations to use when sex chromosome ploidy cutoffs for aneuploidies.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Tuple of ploidy cutoff tuples: ((x_ploidy_cutoffs), (y_ploidy_cutoffs))</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.get_sex_expr">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">get_sex_expr</code><span class="sig-paren">(</span><em class="sig-param">chr_x_ploidy</em>, <em class="sig-param">chr_y_ploidy</em>, <em class="sig-param">x_ploidy_cutoffs</em>, <em class="sig-param">y_ploidy_cutoffs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#get_sex_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.get_sex_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a struct with the following annotation:
- X_karyotype (str)
- Y_karyotype (str)
- sex_karyotype (str)</p>
<p>Note that X0 is currently returned as ‘X’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chr_x_ploidy</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Chromosome X ploidy (or relative ploidy)</p></li>
<li><p><strong>chr_y_ploidy</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>) – Chromosome Y ploidy (or relative ploidy)</p></li>
<li><p><strong>x_ploidy_cutoffs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Tuple of X chromosome ploidy cutoffs: (upper cutoff for single X, (lower cutoff for double X, upper cutoff for double X), lower cutoff for triple X)</p></li>
<li><p><strong>y_ploidy_cutoffs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]) – Tuple of Y chromosome ploidy cutoffs: ((lower cutoff for single Y, upper cutoff for single Y), lower cutoff for double Y)</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Struct containing X_karyotype, Y_karyotype, and sex_karyotype</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.compute_related_samples_to_drop">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">compute_related_samples_to_drop</code><span class="sig-paren">(</span><em class="sig-param">relatedness_ht</em>, <em class="sig-param">rank_ht</em>, <em class="sig-param">kin_threshold</em>, <em class="sig-param">filtered_samples=None</em>, <em class="sig-param">min_related_hard_filter=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#compute_related_samples_to_drop"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.compute_related_samples_to_drop" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p><cite>relatedness_ht</cite> should be keyed by exactly two fields of the same type, identifying the pair of samples for each row.</p></li>
<li><p><cite>rank_ht</cite> should be keyed by a single key of the same type as a single sample identifier in <cite>relatedness_ht</cite>.</p></li>
</ul>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>relatedness_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – relatedness HT, as produced by e.g. pc-relate</p></li>
<li><p><strong>kin_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Kinship threshold to consider two samples as related</p></li>
<li><p><strong>rank_ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Table with a global rank for each sample (smaller is preferred)</p></li>
<li><p><strong>filtered_samples</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.SetExpression.html#hail.expr.SetExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SetExpression</span></code></a>]) – An optional set of samples to exclude (e.g. these samples were hard-filtered)  These samples will then appear in the resulting samples to drop.</p></li>
<li><p><strong>min_related_hard_filter</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – If provided, any sample that is related to more samples than this parameter will be filtered prior to computing the maximal independent set and appear in the results.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A Table with the list of the samples to drop along with their rank.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.run_pca_with_relateds">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">run_pca_with_relateds</code><span class="sig-paren">(</span><em class="sig-param">qc_mt</em>, <em class="sig-param">related_samples_to_drop</em>, <em class="sig-param">n_pcs=10</em>, <em class="sig-param">autosomes_only=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#run_pca_with_relateds"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.run_pca_with_relateds" title="Permalink to this definition">¶</a></dt>
<dd><p>First runs PCA excluding the given related samples,
then projects these samples in the PC space to return scores for all samples.</p>
<p>The <cite>related_samples_to_drop</cite> Table has to be keyed by the sample ID and all samples present in this
table will be excluded from the PCA.</p>
<p>The loadings Table returned also contains a <cite>pca_af</cite> annotation which is the allele frequency
used for PCA. This is useful to project other samples in the PC space.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>qc_mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input QC MT</p></li>
<li><p><strong>related_samples_to_drop</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]) – Optional table of related samples to drop</p></li>
<li><p><strong>n_pcs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>) – Number of PCs to compute</p></li>
<li><p><strong>autosomes_only</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to run the analysis on autosomes only</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>], <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>, <a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>eigenvalues, scores and loadings</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.compute_stratified_metrics_filter">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">compute_stratified_metrics_filter</code><span class="sig-paren">(</span><em class="sig-param">ht</em>, <em class="sig-param">qc_metrics</em>, <em class="sig-param">strata=None</em>, <em class="sig-param">lower_threshold=4.0</em>, <em class="sig-param">upper_threshold=4.0</em>, <em class="sig-param">metric_threshold=None</em>, <em class="sig-param">filter_name='qc_metrics_filters'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#compute_stratified_metrics_filter"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.compute_stratified_metrics_filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute median, MAD, and upper and lower thresholds for each metric used in outlier filtering</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – HT containing relevant sample QC metric annotations</p></li>
<li><p><strong>qc_metrics</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – list of metrics (name and expr) for which to compute the critical values for filtering outliers</p></li>
<li><p><strong>strata</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.Expression.html#hail.expr.Expression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Expression</span></code></a>]]) – List of annotations used for stratification. These metrics should be discrete types!</p></li>
<li><p><strong>lower_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Lower MAD threshold</p></li>
<li><p><strong>upper_threshold</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>) – Upper MAD threshold</p></li>
<li><p><strong>metric_threshold</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]]) – Can be used to specify different (lower, upper) thresholds for one or more metrics</p></li>
<li><p><strong>filter_name</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>) – Name of resulting filters annotation</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Table grouped by strata, with upper and lower threshold values computed for each sample QC metric</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.compute_qc_metrics_residuals">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">compute_qc_metrics_residuals</code><span class="sig-paren">(</span><em class="sig-param">ht</em>, <em class="sig-param">pc_scores</em>, <em class="sig-param">qc_metrics</em>, <em class="sig-param">use_pc_square=True</em>, <em class="sig-param">n_pcs=None</em>, <em class="sig-param">regression_sample_inclusion_expr=&lt;BooleanExpression of type bool&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#compute_qc_metrics_residuals"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.compute_qc_metrics_residuals" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes QC metrics residuals after regressing out PCs (and optionally PC^2)</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <cite>regression_sample_inclusion_expr</cite> can be used to select a subset of the samples to include in the regression calculation.
Residuals are always computed for all samples.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ht</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a>) – Input sample QC metrics HT</p></li>
<li><p><strong>pc_scores</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.ArrayNumericExpression.html#hail.expr.ArrayNumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">ArrayNumericExpression</span></code></a>) – The expression in the input HT that stores the PC scores</p></li>
<li><p><strong>qc_metrics</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.NumericExpression.html#hail.expr.NumericExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">NumericExpression</span></code></a>]) – A dictionary with the name of each QC metric to compute residuals for and their corresponding expression in the input HT.</p></li>
<li><p><strong>use_pc_square</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">bool</span></code>) – Whether to  use PC^2 in the regression or not</p></li>
<li><p><strong>n_pcs</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]) – Numer of PCs to use. If not set, then all PCs in <cite>pc_scores</cite> are used.</p></li>
<li><p><strong>regression_sample_inclusion_expr</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>) – An optional expression to select samples to include in the regression calculation.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Table with QC metrics residuals</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.add_filters_expr">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">add_filters_expr</code><span class="sig-paren">(</span><em class="sig-param">filters</em>, <em class="sig-param">current_filters=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#add_filters_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.add_filters_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an expression to create or add filters.
For each entry in the <cite>filters</cite> dictionary, if the value evaluates to <cite>True</cite>,
then the key is added as a filter name.</p>
<p>Current filters are kept if provided using <cite>current_filters</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filters</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – The filters and their expressions</p></li>
<li><p><strong>current_filters</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.SetExpression.html#hail.expr.SetExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SetExpression</span></code></a>]) – The set of current filters</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.SetExpression.html#hail.expr.SetExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">SetExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>An expression that can be used to annotate the filters</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.merge_sample_qc_expr">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">merge_sample_qc_expr</code><span class="sig-paren">(</span><em class="sig-param">sample_qc_exprs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#merge_sample_qc_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.merge_sample_qc_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates an expression that merges results from non-overlapping strata of hail.sample_qc</p>
<p>E.g.:</p>
<ul class="simple">
<li><p>Compute autosomes and sex chromosomes metrics separately, then merge results</p></li>
<li><p>Compute bi-allelic and multi-allelic metrics separately, then merge results</p></li>
</ul>
<p>Note regarding the merging of <code class="docutils literal notranslate"><span class="pre">dp_stats</span></code> and <code class="docutils literal notranslate"><span class="pre">gq_stats</span></code>:
Because <code class="docutils literal notranslate"><span class="pre">n</span></code> is needed to aggregate <code class="docutils literal notranslate"><span class="pre">stdev</span></code>, <code class="docutils literal notranslate"><span class="pre">n_called</span></code> is used for this purpose.
This should work very well on a standard GATK VCF and it essentially assumes that:</p>
<ul class="simple">
<li><p>samples that are called have <cite>DP</cite> and <cite>GQ</cite> fields</p></li>
<li><p>samples that are not called do not have <cite>DP</cite> and <cite>GQ</cite> fields</p></li>
</ul>
<p>Even if these assumptions are broken for some genotypes, it shouldn’t matter too much.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>sample_qc_exprs</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a>]) – List of sample QC struct expressions for each stratification</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.expr.StructExpression.html#hail.expr.StructExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">StructExpression</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Combined sample QC results</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="gnomad.utils.sample_qc.compute_stratified_sample_qc">
<code class="sig-prename descclassname">gnomad.utils.sample_qc.</code><code class="sig-name descname">compute_stratified_sample_qc</code><span class="sig-paren">(</span><em class="sig-param">mt</em>, <em class="sig-param">strata</em>, <em class="sig-param">tmp_ht_prefix</em>, <em class="sig-param">gt_expr</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/gnomad/utils/sample_qc.html#compute_stratified_sample_qc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#gnomad.utils.sample_qc.compute_stratified_sample_qc" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs hl.sample_qc on different strata and then also merge the results into a single expression.
Note that strata should be non-overlapping,  e.g. SNV vs indels or bi-allelic vs multi-allelic</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mt</strong> (<a class="reference external" href="https://hail.is/docs/0.2/hail.MatrixTable.html#hail.MatrixTable" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">MatrixTable</span></code></a>) – Input MT</p></li>
<li><p><strong>strata</strong> (<code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <a class="reference external" href="https://hail.is/docs/0.2/hail.expr.BooleanExpression.html#hail.expr.BooleanExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">BooleanExpression</span></code></a>]) – Strata names and filtering expressions</p></li>
<li><p><strong>tmp_ht_prefix</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]) – Optional path prefix to write the intermediate strata results to (recommended for larger datasets)</p></li>
<li><p><strong>gt_expr</strong> (<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference external" href="https://hail.is/docs/0.2/hail.expr.CallExpression.html#hail.expr.CallExpression" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">CallExpression</span></code></a>]) – Optional entry field storing the genotype (if not specified, then it is assumed that it is stored in mt.GT)</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://hail.is/docs/0.2/hail.Table.html#hail.Table" title="(in Hail v0.2)"><code class="xref py py-class docutils literal notranslate"><span class="pre">Table</span></code></a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Sample QC table, including strat-specific numbers</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="slack.html" class="btn btn-neutral float-right" title="gnomad.utils.slack" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="rf.html" class="btn btn-neutral float-left" title="gnomad.utils.rf" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>