

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gnomad.utils.sample_qc &mdash; gnomad master documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> gnomad
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference/index.html">API Reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>gnomad.utils.sample_qc</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for gnomad.utils.sample_qc</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">.generic</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">.gnomad_functions</span> <span class="kn">import</span> <span class="n">logger</span><span class="p">,</span> <span class="n">filter_to_adj</span>
<span class="kn">from</span> <span class="nn">.sparse_mt</span> <span class="kn">import</span> <span class="n">impute_sex_ploidy</span>
<span class="kn">from</span> <span class="nn">gnomad.utils.relatedness</span> <span class="kn">import</span> <span class="n">get_duplicated_samples</span>


<div class="viewcode-block" id="filter_rows_for_qc"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.filter_rows_for_qc">[docs]</a><span class="k">def</span> <span class="nf">filter_rows_for_qc</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">min_af</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
        <span class="n">min_callrate</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
        <span class="n">min_inbreeding_coeff_threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.8</span><span class="p">,</span>
        <span class="n">min_hardy_weinberg_threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1e-8</span><span class="p">,</span>
        <span class="n">apply_hard_filters</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">bi_allelic_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">snv_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Annotates rows with `sites_callrate`, `site_inbreeding_coeff` and `af`, then applies thresholds.</span>
<span class="sd">    AF and callrate thresholds are taken from gnomAD QC; inbreeding coeff, MQ, FS and QD filters are taken from GATK best practices</span>

<span class="sd">    .. note::</span>

<span class="sd">        This function expect the typical ``info`` annotation of type struct with fields ``MQ``, ``FS`` and ``QD``</span>
<span class="sd">        if applying hard filters.</span>

<span class="sd">    :param mt: Input MT</span>
<span class="sd">    :param min_af: Minimum site AF to keep. Not applied if set to ``None``.</span>
<span class="sd">    :param min_callrate: Minimum site call rate to keep. Not applied if set to ``None``.</span>
<span class="sd">    :param min_inbreeding_coeff_threshold: Minimum site inbreeding coefficient to keep. Not applied if set to ``None``.</span>
<span class="sd">    :param min_hardy_weinberg_threshold: Minimum site HW test p-value to keep. Not applied if set to ``None``.</span>
<span class="sd">    :paramapply_hard_filters: Whether to apply standard GAKT default site hard filters: QD &gt;= 2, FS &lt;= 60 and MQ &gt;= 30</span>
<span class="sd">    :parambi_allelic_only: Whether to only keep bi-allelic sites or include multi-allelic sites too</span>
<span class="sd">    :paramsnv_only: Whether to only keep SNVs or include other variant types</span>
<span class="sd">    :return: annotated and filtered table</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">annotation_expr</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="k">if</span> <span class="n">min_af</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_expr</span><span class="p">[</span><span class="s1">&#39;af&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span> <span class="o">/</span> <span class="mi">2</span>
    <span class="k">if</span> <span class="n">min_callrate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_expr</span><span class="p">[</span><span class="s1">&#39;site_callrate&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">fraction</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">min_inbreeding_coeff_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_expr</span><span class="p">[</span><span class="s1">&#39;site_inbreeding_coeff&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bi_allelic_site_inbreeding_expr</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">min_hardy_weinberg_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">annotation_expr</span><span class="p">[</span><span class="s1">&#39;hwe&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">hardy_weinberg_test</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">GT</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">annotation_expr</span><span class="p">:</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="o">**</span><span class="n">annotation_expr</span><span class="p">)</span>

    <span class="n">filter_expr</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">min_af</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">af</span> <span class="o">&gt;</span> <span class="n">min_af</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">min_callrate</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">site_callrate</span> <span class="o">&gt;</span> <span class="n">min_callrate</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">min_inbreeding_coeff_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">site_inbreeding_coeff</span> <span class="o">&gt;</span> <span class="n">min_inbreeding_coeff_threshold</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">min_hardy_weinberg_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">hwe</span><span class="o">.</span><span class="n">p_value</span> <span class="o">&gt;</span> <span class="n">min_hardy_weinberg_threshold</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">snv_only</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="k">if</span> <span class="n">bi_allelic_only</span><span class="p">:</span>
        <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">bi_allelic_expr</span><span class="p">(</span><span class="n">mt</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">apply_hard_filters</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;info&#39;</span> <span class="ow">in</span> <span class="n">mt</span><span class="o">.</span><span class="n">row_value</span><span class="p">:</span>
            <span class="k">if</span> <span class="s1">&#39;QD&#39;</span> <span class="ow">in</span> <span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
                <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">QD</span> <span class="o">&gt;=</span> <span class="mi">2</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Could not apply QD hard filter, as `info.QD` not found in schema.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;FS&#39;</span> <span class="ow">in</span> <span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
                <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">FS</span> <span class="o">&lt;=</span> <span class="mi">60</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Could not apply FS hard filter, as `info.FS` not found in schema.&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="s1">&#39;MQ&#39;</span> <span class="ow">in</span> <span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
                <span class="n">filter_expr</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mt</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">MQ</span> <span class="o">&gt;=</span> <span class="mi">30</span><span class="p">))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Could not apply MQ hard filter, as `info.MQ` not found in schema.&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Could not apply hard filters as `info` not found in schema.&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">functools</span><span class="o">.</span><span class="n">reduce</span><span class="p">(</span><span class="n">operator</span><span class="o">.</span><span class="n">iand</span><span class="p">,</span> <span class="n">filter_expr</span><span class="p">))</span></div>


<div class="viewcode-block" id="get_qc_mt"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.get_qc_mt">[docs]</a><span class="k">def</span> <span class="nf">get_qc_mt</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">adj_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">min_af</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
        <span class="n">min_callrate</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.99</span><span class="p">,</span>
        <span class="n">min_inbreeding_coeff_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="o">-</span><span class="mf">0.8</span><span class="p">,</span>
        <span class="n">min_hardy_weinberg_threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">1e-8</span><span class="p">,</span>
        <span class="n">apply_hard_filters</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">ld_r2</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.1</span><span class="p">,</span>
        <span class="n">filter_lcr</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">filter_decoy</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">filter_segdup</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">filter_exome_low_coverage_regions</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">high_conf_regions</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a QC-ready MT by keeping:</span>

<span class="sd">    - Variants outside known problematic regions</span>
<span class="sd">    - Bi-allelic SNVs only</span>
<span class="sd">    - Variants passing hard thresholds</span>
<span class="sd">    - Variants passing the set call rate and MAF thresholds</span>
<span class="sd">    - Genotypes passing on gnomAD ADJ criteria (GQ&gt;=20, DP&gt;=10, AB&gt;0.2 for hets)</span>

<span class="sd">    In addition, the MT will be LD-pruned if `ld_r2` is set.</span>

<span class="sd">    :param mt: Input MT</span>
<span class="sd">    :param adj_only: If set, only ADJ genotypes are kept. This filter is applied before the call rate and AF calculation.</span>
<span class="sd">    :param min_af: Minimum allele frequency to keep</span>
<span class="sd">    :param min_callrate: Minimum call rate to keep</span>
<span class="sd">    :param min_inbreeding_coeff_threshold: Minimum site inbreeding coefficient to keep</span>
<span class="sd">    :param min_hardy_weinberg_threshold: Minimum site HW test p-value to keep</span>
<span class="sd">    :param apply_hard_filters: Whether to apply standard GAKT default site hard filters: QD &gt;= 2, FS &lt;= 60 and MQ &gt;= 30</span>
<span class="sd">    :param ld_r2: Minimum r2 to keep when LD-pruning (set to `None` for no LD pruning)</span>
<span class="sd">    :param filter_lcr: Filter LCR regions</span>
<span class="sd">    :param filter_decoy: Filter decoy regions</span>
<span class="sd">    :param filter_segdup: Filter segmental duplication regions</span>
<span class="sd">    :param filter_exome_low_coverage_regions: If set, only high coverage exome regions (computed from gnomAD are kept)</span>
<span class="sd">    :param high_conf_regions: If given, the data will be filtered to only include variants in those regions</span>
<span class="sd">    :return: Filtered MT</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creating QC MatrixTable&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">ld_r2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;The LD-prune step of this function requires non-preemptible workers only!&quot;</span><span class="p">)</span>

    <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">filter_low_conf_regions</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">,</span>
        <span class="n">filter_lcr</span><span class="o">=</span><span class="n">filter_lcr</span><span class="p">,</span>
        <span class="n">filter_decoy</span><span class="o">=</span><span class="n">filter_decoy</span><span class="p">,</span>
        <span class="n">filter_segdup</span><span class="o">=</span><span class="n">filter_segdup</span><span class="p">,</span>
        <span class="n">filter_exome_low_coverage_regions</span><span class="o">=</span><span class="n">filter_exome_low_coverage_regions</span><span class="p">,</span> <span class="n">high_conf_regions</span><span class="o">=</span><span class="n">high_conf_regions</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">adj_only</span><span class="p">:</span>
        <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">filter_to_adj</span><span class="p">(</span><span class="n">qc_mt</span><span class="p">)</span>  <span class="c1"># TODO: Make sure that this works fine before call rate filtering</span>

    <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">filter_rows_for_qc</span><span class="p">(</span>
        <span class="n">qc_mt</span><span class="p">,</span>
        <span class="n">min_af</span><span class="p">,</span>
        <span class="n">min_callrate</span><span class="p">,</span>
        <span class="n">min_inbreeding_coeff_threshold</span><span class="p">,</span>
        <span class="n">min_hardy_weinberg_threshold</span><span class="p">,</span>
        <span class="n">apply_hard_filters</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">ld_r2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="n">unfiltered_qc_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">unfilter_entries</span><span class="p">()</span>
        <span class="n">pruned_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">ld_prune</span><span class="p">(</span><span class="n">unfiltered_qc_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span> <span class="n">r2</span><span class="o">=</span><span class="n">ld_r2</span><span class="p">)</span>
        <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">pruned_ht</span><span class="p">[</span><span class="n">qc_mt</span><span class="o">.</span><span class="n">row_key</span><span class="p">]))</span>

    <span class="n">qc_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">qc_mt_params</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">adj_only</span><span class="o">=</span><span class="n">adj_only</span><span class="p">,</span>
            <span class="n">min_af</span><span class="o">=</span><span class="n">min_af</span><span class="p">,</span>
            <span class="n">min_callrate</span><span class="o">=</span><span class="n">min_callrate</span><span class="p">,</span>
            <span class="n">inbreeding_coeff_threshold</span><span class="o">=</span><span class="n">min_inbreeding_coeff_threshold</span><span class="p">,</span>
            <span class="n">apply_hard_filters</span><span class="o">=</span><span class="n">apply_hard_filters</span><span class="p">,</span>
            <span class="n">ld_r2</span><span class="o">=</span><span class="n">ld_r2</span> <span class="k">if</span> <span class="n">ld_r2</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">hl</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tfloat32</span><span class="p">),</span>
            <span class="n">filter_exome_low_coverage_regions</span><span class="o">=</span><span class="n">filter_exome_low_coverage_regions</span><span class="p">,</span>
            <span class="n">high_conf_regions</span><span class="o">=</span><span class="n">high_conf_regions</span> <span class="k">if</span> <span class="n">high_conf_regions</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">hl</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tarray</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">))</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">annotate_cols</span><span class="p">(</span><span class="n">sample_callrate</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">fraction</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">qc_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">)))</span></div>


<div class="viewcode-block" id="compute_callrate_mt"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.compute_callrate_mt">[docs]</a><span class="k">def</span> <span class="nf">compute_callrate_mt</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">intervals_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">bi_allelic_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">autosomes_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">match</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes a sample/interval MT with each entry containing the call rate for that sample/interval.</span>
<span class="sd">    This can be used as input for imputing exome sequencing platforms.</span>

<span class="sd">    .. note::</span>

<span class="sd">        The input interval HT should have a key of type Interval.</span>
<span class="sd">        The resulting table will have a key of the same type as the `intervals_ht` table and</span>
<span class="sd">        contain an `interval_info` field containing all non-key fields of the `intervals_ht`.</span>

<span class="sd">    :param mt: Input MT</span>
<span class="sd">    :param intervals_ht: Table containing the intervals. This table has to be keyed by locus.</span>
<span class="sd">    :param bi_allelic_only: If set, only bi-allelic sites are used for the computation</span>
<span class="sd">    :param autosomes_only: If set, only autosomal intervals are used.</span>
<span class="sd">    :param matches: If set, returns all intervals in intervals_ht that overlap the locus in the input MT.</span>
<span class="sd">    :return: Callrate MT</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Computing call rate MatrixTable&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">intervals_ht</span><span class="o">.</span><span class="n">key</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">1</span> <span class="ow">or</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">intervals_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">IntervalExpression</span><span class="p">):</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Call rate matrix computation expects `intervals_ht` with a key of type Interval. Found: </span><span class="si">{intervals_ht.key}</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">autosomes_only</span><span class="p">:</span>
        <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">filter_to_autosomes</span><span class="p">(</span><span class="n">mt</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">bi_allelic_only</span><span class="p">:</span>
        <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">bi_allelic_expr</span><span class="p">(</span><span class="n">callrate_mt</span><span class="p">))</span>

    <span class="n">intervals_ht</span> <span class="o">=</span> <span class="n">intervals_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">_interval_key</span><span class="o">=</span><span class="n">intervals_ht</span><span class="o">.</span><span class="n">key</span><span class="p">)</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">_interval_key</span><span class="o">=</span><span class="n">intervals_ht</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">locus</span><span class="p">,</span> <span class="n">all_matches</span><span class="o">=</span><span class="n">match</span><span class="p">)</span><span class="o">.</span><span class="n">_interval_key</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
        <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">explode_rows</span><span class="p">(</span><span class="s1">&#39;_interval_key&#39;</span><span class="p">)</span>

    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">_interval_key</span><span class="o">.</span><span class="n">interval</span><span class="p">))</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="n">GT</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">),</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">()))</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">group_rows_by</span><span class="p">(</span><span class="o">**</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">_interval_key</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">callrate</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">fraction</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">)))</span>
    <span class="n">intervals_ht</span> <span class="o">=</span> <span class="n">intervals_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;_interval_key&#39;</span><span class="p">)</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">interval_info</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="o">**</span><span class="n">intervals_ht</span><span class="p">[</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">row_key</span><span class="p">]))</span>
    <span class="k">return</span> <span class="n">callrate_mt</span></div>


<div class="viewcode-block" id="run_platform_pca"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.run_platform_pca">[docs]</a><span class="k">def</span> <span class="nf">run_platform_pca</span><span class="p">(</span>
        <span class="n">callrate_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">binarization_threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">float</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.25</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs a PCA on a sample/interval MT with each entry containing the call rate.</span>
<span class="sd">    When `binzarization_threshold` is set, the callrate is transformed to a 0/1 value based on the threshold.</span>
<span class="sd">    E.g. with the default threshold of 0.25, all entries with a callrate &lt; 0.25 are considered as 0s, others as 1s.</span>

<span class="sd">    :param callrate_mt: Input callrate MT</span>
<span class="sd">    :param binarization_threshold: binzarization_threshold. None is no threshold desired</span>
<span class="sd">    :return: eigenvalues, scores_ht, loadings_ht</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running platform PCA&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">binarization_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span><span class="n">callrate</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">int</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">callrate</span> <span class="o">&gt;</span> <span class="n">binarization_threshold</span><span class="p">))</span>
    <span class="c1"># Center until Hail&#39;s PCA does it for you</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">mean_callrate</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">callrate</span><span class="p">))</span>
    <span class="n">callrate_mt</span> <span class="o">=</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">annotate_entries</span><span class="p">(</span><span class="n">callrate</span><span class="o">=</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">callrate</span> <span class="o">-</span> <span class="n">callrate_mt</span><span class="o">.</span><span class="n">mean_callrate</span><span class="p">)</span>
    <span class="n">eigenvalues</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">loadings</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">callrate_mt</span><span class="o">.</span><span class="n">callrate</span><span class="p">,</span> <span class="n">compute_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  <span class="c1"># TODO:  Evaluate whether computing loadings is a good / worthy thing</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Platform PCA eigenvalues: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">eigenvalues</span><span class="p">))</span>

    <span class="k">return</span> <span class="n">eigenvalues</span><span class="p">,</span> <span class="n">scores</span><span class="p">,</span> <span class="n">loadings</span></div>


<div class="viewcode-block" id="assign_platform_from_pcs"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.assign_platform_from_pcs">[docs]</a><span class="k">def</span> <span class="nf">assign_platform_from_pcs</span><span class="p">(</span>
        <span class="n">platform_pca_scores_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">pc_scores_ann</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;scores&#39;</span><span class="p">,</span>
        <span class="n">hdbscan_min_cluster_size</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">hdbscan_min_samples</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Assigns platforms using HBDSCAN on the results of call rate PCA.</span>

<span class="sd">    :param platform_pca_scores_ht: Input table with the PCA score for each sample</span>
<span class="sd">    :param pc_scores_ann: Field containing the scores</span>
<span class="sd">    :param hdbscan_min_cluster_size: HDBSCAN `min_cluster_size` parameter. If not specified the smallest of 500 and 0.1*n_samples will be used.</span>
<span class="sd">    :param hdbscan_min_samples: HDBSCAN `min_samples` parameter</span>
<span class="sd">    :return: A Table with a `qc_platform` annotation containing the platform based on HDBSCAN clustering</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">hdbscan</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Assigning platforms based on platform PCA clustering&quot;</span><span class="p">)</span>

    <span class="c1"># Read and format data for clustering</span>
    <span class="n">data</span> <span class="o">=</span> <span class="n">platform_pca_scores_ht</span><span class="o">.</span><span class="n">to_pandas</span><span class="p">()</span>
    <span class="n">callrate_data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="n">pc_scores_ann</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Assigning platforms to </span><span class="si">{}</span><span class="s1"> samples.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">callrate_data</span><span class="p">)))</span>

    <span class="c1"># Cluster data</span>
    <span class="k">if</span> <span class="n">hdbscan_min_cluster_size</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">hdbscan_min_cluster_size</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="mf">0.1</span> <span class="o">*</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">clusterer</span> <span class="o">=</span> <span class="n">hdbscan</span><span class="o">.</span><span class="n">HDBSCAN</span><span class="p">(</span><span class="n">min_cluster_size</span><span class="o">=</span><span class="n">hdbscan_min_cluster_size</span><span class="p">,</span> <span class="n">min_samples</span><span class="o">=</span><span class="n">hdbscan_min_samples</span><span class="p">)</span>
    <span class="n">cluster_labels</span> <span class="o">=</span> <span class="n">clusterer</span><span class="o">.</span><span class="n">fit_predict</span><span class="p">(</span><span class="n">callrate_data</span><span class="p">)</span>
    <span class="n">n_clusters</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">cluster_labels</span><span class="p">))</span> <span class="o">-</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span> <span class="ow">in</span> <span class="n">cluster_labels</span><span class="p">)</span>  <span class="c1"># NOTE: -1 is the label for noisy (un-classifiable) data points</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Found </span><span class="si">{}</span><span class="s1"> unique platforms during platform imputation.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">n_clusters</span><span class="p">))</span>

    <span class="n">data</span><span class="p">[</span><span class="s1">&#39;qc_platform&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">cluster_labels</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">from_pandas</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="p">[</span><span class="o">*</span><span class="n">platform_pca_scores_ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">qc_platform</span><span class="o">=</span><span class="s1">&#39;platform_&#39;</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">str</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">qc_platform</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="default_annotate_sex"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.default_annotate_sex">[docs]</a><span class="k">def</span> <span class="nf">default_annotate_sex</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">is_sparse</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">excluded_intervals</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">included_intervals</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">normalization_contig</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;chr20&#39;</span><span class="p">,</span>
        <span class="n">sites_ht</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">aaf_expr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">gt_expr</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;GT&#39;</span><span class="p">,</span>
        <span class="n">f_stat_cutoff</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
        <span class="n">aaf_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Imputes sample sex based on X-chromosome heterozygosity and sex chromosome ploidy.</span>
<span class="sd"> </span>
<span class="sd">    Returns Table with the following fields:</span>
<span class="sd">        - s (str): Sample</span>
<span class="sd">        - chr20_mean_dp (float32): Sample&#39;s mean coverage over chromosome 20.</span>
<span class="sd">        - chrX_mean_dp (float32): Sample&#39;s mean coverage over chromosome X.</span>
<span class="sd">        - chrY_mean_dp (float32): Sample&#39;s mean coverage over chromosome Y.</span>
<span class="sd">        - chrX_ploidy (float32): Sample&#39;s imputed ploidy over chromosome X.</span>
<span class="sd">        - chrY_ploidy (float32): Sample&#39;s imputed ploidy over chromosome Y.</span>
<span class="sd">        - f_stat (float64): Sample f-stat. Calculated using hl.impute_sex.</span>
<span class="sd">        - n_called (int64): Number of variants with a genotype call. Calculated using hl.impute_sex.</span>
<span class="sd">        - expected_homs (float64): Expected number of homozygotes. Calculated using hl.impute_sex.</span>
<span class="sd">        - observed_homs (int64): Expected number of homozygotes. Calculated using hl.impute_sex.</span>
<span class="sd">        - X_karyotype (str): Sample&#39;s chromosome X karyotype.</span>
<span class="sd">        - Y_karyotype (str): Sample&#39;s chromosome Y karyotype.</span>
<span class="sd">        - sex_karyotype (str): Sample&#39;s sex karyotype.</span>

<span class="sd">    :param mt: Input MatrixTable</span>
<span class="sd">    :param bool is_sparse: Whether input MatrixTable is in sparse data format</span>
<span class="sd">    :param excluded_intervals: Optional table of intervals to exclude from the computation.</span>
<span class="sd">    :param included_intervals: Optional table of intervals to use in the computation. REQUIRED for exomes.</span>
<span class="sd">    :param normalization_contig: Which chromosome to use to normalize sex chromosome coverage. Used in determining sex chromosome ploidies.</span>
<span class="sd">    :param sites_ht: Optional Table to use. If present, filters input MatrixTable to sites in this Table prior to imputing sex,</span>
<span class="sd">                    and pulls alternate allele frequency from this Table.</span>
<span class="sd">    :param aaf_expr: Optional. Name of field in input MatrixTable with alternate allele frequency.</span>
<span class="sd">    :param gt_expr: Name of entry field storing the genotype. Default: &#39;GT&#39;</span>
<span class="sd">    :param f_stat_cutoff: f-stat to roughly divide &#39;XX&#39; from &#39;XY&#39; samples. Assumes XX samples are below cutoff and XY are above cutoff.</span>
<span class="sd">    :param float aaf_threshold: Minimum alternate allele frequency to be used in f-stat calculations.</span>
<span class="sd">    :return: Table of samples and their imputed sex karyotypes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Imputing sex chromosome ploidies...&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">is_sparse</span><span class="p">:</span>
        <span class="n">ploidy_ht</span> <span class="o">=</span> <span class="n">impute_sex_ploidy</span><span class="p">(</span>
                <span class="n">mt</span><span class="p">,</span> <span class="n">excluded_intervals</span><span class="p">,</span> <span class="n">included_intervals</span><span class="p">,</span>
                <span class="n">normalization_contig</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span><span class="p">(</span><span class="s2">&quot;Imputing sex ploidy does not exist yet for dense data.&quot;</span><span class="p">)</span>

    <span class="n">x_contigs</span> <span class="o">=</span> <span class="n">get_reference_genome</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">locus</span><span class="p">)</span><span class="o">.</span><span class="n">x_contigs</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtering mt to biallelic SNPs in X contigs: </span><span class="si">{x_contigs}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="s1">&#39;was_split&#39;</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">row</span><span class="p">):</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">((</span><span class="o">~</span><span class="n">mt</span><span class="o">.</span><span class="n">was_split</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">((</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">mt</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">filter_intervals</span><span class="p">(</span><span class="n">mt</span><span class="p">,</span> <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">parse_locus_interval</span><span class="p">(</span><span class="n">contig</span><span class="p">)</span> <span class="k">for</span> <span class="n">contig</span> <span class="ow">in</span> <span class="n">x_contigs</span><span class="p">])</span>

    <span class="k">if</span> <span class="n">sites_ht</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">aaf_expr</span> <span class="o">==</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;sites_ht was provided, but aaf_expr is missing. Assuming name of field with alternate allele frequency is &#39;AF&#39;.&quot;</span><span class="p">)</span>
            <span class="n">aaf_expr</span> <span class="o">=</span> <span class="s2">&quot;AF&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Filtering to provided sites&quot;</span><span class="p">)</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="o">**</span><span class="n">sites_ht</span><span class="p">[</span><span class="n">mt</span><span class="o">.</span><span class="n">row_key</span><span class="p">])</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">mt</span><span class="p">[</span><span class="n">aaf_expr</span><span class="p">]))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating inbreeding coefficient on chrX&quot;</span><span class="p">)</span>
    <span class="n">sex_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">impute_sex</span><span class="p">(</span><span class="n">mt</span><span class="p">[</span><span class="n">gt_expr</span><span class="p">],</span> <span class="n">aaf_threshold</span><span class="o">=</span><span class="n">aaf_threshold</span><span class="p">,</span> <span class="n">male_threshold</span><span class="o">=</span><span class="n">f_stat_cutoff</span><span class="p">,</span> <span class="n">female_threshold</span><span class="o">=</span><span class="n">f_stat_cutoff</span><span class="p">,</span> <span class="n">aaf</span><span class="o">=</span><span class="n">aaf_expr</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating sex ht with sex chromosome ploidies&quot;</span><span class="p">)</span>
    <span class="n">sex_ht</span> <span class="o">=</span> <span class="n">sex_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="n">ploidy_ht</span><span class="p">[</span><span class="n">sex_ht</span><span class="o">.</span><span class="n">key</span><span class="p">])</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Inferring sex karyotypes&quot;</span><span class="p">)</span>
    <span class="n">x_ploidy_cutoffs</span><span class="p">,</span> <span class="n">y_ploidy_cutoffs</span> <span class="o">=</span> <span class="n">get_ploidy_cutoffs</span><span class="p">(</span><span class="n">sex_ht</span><span class="p">,</span> <span class="n">f_stat_cutoff</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">sex_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="o">**</span><span class="n">get_sex_expr</span><span class="p">(</span>
                <span class="n">sex_ht</span><span class="o">.</span><span class="n">chrX_ploidy</span><span class="p">,</span>
                <span class="n">sex_ht</span><span class="o">.</span><span class="n">chrY_ploidy</span><span class="p">,</span>
                <span class="n">x_ploidy_cutoffs</span><span class="p">,</span>
                <span class="n">y_ploidy_cutoffs</span>
        <span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="get_ploidy_cutoffs"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.get_ploidy_cutoffs">[docs]</a><span class="k">def</span> <span class="nf">get_ploidy_cutoffs</span><span class="p">(</span>
    <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> 
    <span class="n">f_stat_cutoff</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span> 
    <span class="n">normal_ploidy_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">5</span><span class="p">,</span> 
    <span class="n">aneuploidy_cutoff</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">6</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="nb">float</span><span class="p">],</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="nb">float</span><span class="p">]]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Gets chromosome X and Y ploidy cutoffs for XY and XX samples. Note this assumes the input hail Table has the fields f_stat, chrX_ploidy, and chrY_ploidy.</span>
<span class="sd">    Returns a tuple of sex chromosome ploidy cutoffs: ((x_ploidy_cutoffs), (y_ploidy_cutoffs)).</span>
<span class="sd">    x_ploidy_cutoffs: (upper cutoff for single X, (lower cutoff for double X, upper cutoff for double X), lower cutoff for triple X)</span>
<span class="sd">    y_ploidy_cutoffs: ((lower cutoff for single Y, upper cutoff for single Y), lower cutoff for double Y)</span>

<span class="sd">    Uses the normal_ploidy_cutoff parameter to determine the ploidy cutoffs for XX and XY karyotypes.</span>
<span class="sd">    Uses the aneuploidy_cutoff parameter to determine the cutoffs for sex aneuploidies.</span>

<span class="sd">    Note that f-stat is used only to split the samples into roughly &#39;XX&#39; and &#39;XY&#39; categories and is not used in the final karyotype annotation.</span>

<span class="sd">    :param ht: Table with f_stat and sex chromosome ploidies</span>
<span class="sd">    :param f_stat_cutoff: f-stat to roughly divide &#39;XX&#39; from &#39;XY&#39; samples. Assumes XX samples are below cutoff and XY are above cutoff.</span>
<span class="sd">    :param normal_ploidy_cutoff: Number of standard deviations to use when determining sex chromosome ploidy cutoffs for XX, XY karyotypes.</span>
<span class="sd">    :param aneuploidy_cutoff: Number of standard deviations to use when sex chromosome ploidy cutoffs for aneuploidies.</span>
<span class="sd">    :return: Tuple of ploidy cutoff tuples: ((x_ploidy_cutoffs), (y_ploidy_cutoffs))</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Group sex chromosome ploidy table by f_stat cutoff and get mean/stdev for chrX/Y ploidies</span>
    <span class="n">sex_stats</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">cond</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">f_stat</span> <span class="o">&lt;</span> <span class="n">f_stat_cutoff</span><span class="p">,</span> <span class="s2">&quot;xx&quot;</span><span class="p">,</span> <span class="s2">&quot;xy&quot;</span><span class="p">),</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">x</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">chrX_ploidy</span><span class="p">),</span>
                    <span class="n">y</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">stats</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">chrY_ploidy</span><span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>
    <span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;XX stats: </span><span class="si">{sex_stats[&#39;xx&#39;]}</span><span class="s2">&quot;</span><span class="p">)</span> 
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;XY stats: </span><span class="si">{sex_stats[&#39;xy&#39;]}</span><span class="s2">&quot;</span><span class="p">)</span>

    <span class="n">cutoffs</span> <span class="o">=</span> <span class="p">(</span>
                <span class="p">(</span>
                    <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">normal_ploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">stdev</span><span class="p">),</span>
                    <span class="p">(</span><span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">mean</span> <span class="o">-</span> <span class="p">(</span><span class="n">normal_ploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">stdev</span><span class="p">),</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">normal_ploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">stdev</span><span class="p">)),</span>
                    <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">aneuploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">x</span><span class="o">.</span><span class="n">stdev</span><span class="p">)</span>
                <span class="p">),</span>
                <span class="p">(</span>
                    <span class="p">(</span><span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">normal_ploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">stdev</span><span class="p">),</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">normal_ploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">stdev</span><span class="p">)),</span> 
                    <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">mean</span> <span class="o">+</span> <span class="p">(</span><span class="n">aneuploidy_cutoff</span> <span class="o">*</span> <span class="n">sex_stats</span><span class="p">[</span><span class="s2">&quot;xy&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">y</span><span class="o">.</span><span class="n">stdev</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
 
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;X ploidy cutoffs: </span><span class="si">{cutoffs[0]}</span><span class="s2">&quot;</span><span class="p">)</span> 
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Y ploidy cutoffs: </span><span class="si">{cutoffs[1]}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">cutoffs</span></div>


<div class="viewcode-block" id="get_sex_expr"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.get_sex_expr">[docs]</a><span class="k">def</span> <span class="nf">get_sex_expr</span><span class="p">(</span>
    <span class="n">chr_x_ploidy</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">chr_y_ploidy</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
    <span class="n">x_ploidy_cutoffs</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="nb">float</span><span class="p">],</span>
    <span class="n">y_ploidy_cutoffs</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">],</span> <span class="nb">float</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates a struct with the following annotation:</span>
<span class="sd">    - X_karyotype (str)</span>
<span class="sd">    - Y_karyotype (str)</span>
<span class="sd">    - sex_karyotype (str)</span>

<span class="sd">    Note that X0 is currently returned as &#39;X&#39;.</span>

<span class="sd">    :param chr_x_ploidy: Chromosome X ploidy (or relative ploidy)</span>
<span class="sd">    :param chr_y_ploidy: Chromosome Y ploidy (or relative ploidy)</span>
<span class="sd">    :param x_ploidy_cutoffs: Tuple of X chromosome ploidy cutoffs: (upper cutoff for single X, (lower cutoff for double X, upper cutoff for double X), lower cutoff for triple X)</span>
<span class="sd">    :param y_ploidy_cutoffs: Tuple of Y chromosome ploidy cutoffs: ((lower cutoff for single Y, upper cutoff for single Y), lower cutoff for double Y)</span>
<span class="sd">    :return: Struct containing X_karyotype, Y_karyotype, and sex_karyotype</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">sex_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
        <span class="n">X_karyotype</span><span class="o">=</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="n">chr_x_ploidy</span> <span class="o">&lt;</span> <span class="n">x_ploidy_cutoffs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;X&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">((</span><span class="n">chr_x_ploidy</span> <span class="o">&gt;</span> <span class="n">x_ploidy_cutoffs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">&amp;</span>
                    <span class="p">(</span><span class="n">chr_x_ploidy</span> <span class="o">&lt;</span> <span class="n">x_ploidy_cutoffs</span><span class="p">[</span><span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">])),</span>
                    <span class="s2">&quot;XX&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">(</span><span class="n">chr_x_ploidy</span> <span class="o">&gt;=</span> <span class="n">x_ploidy_cutoffs</span><span class="p">[</span><span class="mi">2</span><span class="p">]),</span>
                    <span class="s2">&quot;XXX&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="s2">&quot;Ambiguous&quot;</span><span class="p">)</span>
        <span class="p">),</span>
        <span class="n">Y_karyotype</span><span class="o">=</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">case</span><span class="p">()</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">chr_y_ploidy</span> <span class="o">&lt;</span> <span class="n">y_ploidy_cutoffs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                    <span class="p">((</span><span class="n">chr_y_ploidy</span> <span class="o">&gt;</span> <span class="n">y_ploidy_cutoffs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">&amp;</span>
                    <span class="p">(</span><span class="n">chr_y_ploidy</span> <span class="o">&lt;</span> <span class="n">y_ploidy_cutoffs</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">1</span><span class="p">])),</span>
                    <span class="s2">&quot;Y&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">when</span><span class="p">(</span>
                     <span class="n">chr_y_ploidy</span> <span class="o">&gt;=</span> <span class="n">y_ploidy_cutoffs</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
                    <span class="s2">&quot;YY&quot;</span><span class="p">)</span>
                <span class="o">.</span><span class="n">default</span><span class="p">(</span><span class="s2">&quot;Ambiguous&quot;</span><span class="p">)</span>
            <span class="p">)</span>
    <span class="p">)</span>
    
    <span class="k">return</span> <span class="n">sex_expr</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">sex_karyotype</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">if_else</span><span class="p">(</span>
            <span class="p">(</span><span class="n">sex_expr</span><span class="o">.</span><span class="n">X_karyotype</span> <span class="o">==</span> <span class="s2">&quot;Ambiguous&quot;</span><span class="p">)</span> <span class="o">|</span>  <span class="p">(</span><span class="n">sex_expr</span><span class="o">.</span><span class="n">Y_karyotype</span> <span class="o">==</span> <span class="s2">&quot;Ambiguous&quot;</span><span class="p">),</span>
            <span class="s2">&quot;Ambiguous&quot;</span><span class="p">,</span>
            <span class="n">sex_expr</span><span class="o">.</span><span class="n">X_karyotype</span> <span class="o">+</span> <span class="n">sex_expr</span><span class="o">.</span><span class="n">Y_karyotype</span>
            <span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="compute_related_samples_to_drop"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.compute_related_samples_to_drop">[docs]</a><span class="k">def</span> <span class="nf">compute_related_samples_to_drop</span><span class="p">(</span>
        <span class="n">relatedness_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">rank_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">kin_threshold</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
        <span class="n">filtered_samples</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">min_related_hard_filter</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
 <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes a Table with the list of samples to drop (and their global rank) to get the maximal independent set of unrelated samples.</span>

<span class="sd">    .. note::</span>

<span class="sd">        - `relatedness_ht` should be keyed by exactly two fields of the same type, identifying the pair of samples for each row.</span>
<span class="sd">        - `rank_ht` should be keyed by a single key of the same type as a single sample identifier in `relatedness_ht`.</span>

<span class="sd">    :param relatedness_ht: relatedness HT, as produced by e.g. pc-relate</span>
<span class="sd">    :param kin_threshold: Kinship threshold to consider two samples as related</span>
<span class="sd">    :param rank_ht: Table with a global rank for each sample (smaller is preferred)</span>
<span class="sd">    :param filtered_samples: An optional set of samples to exclude (e.g. these samples were hard-filtered)  These samples will then appear in the resulting samples to drop.</span>
<span class="sd">    :param min_related_hard_filter: If provided, any sample that is related to more samples than this parameter will be filtered prior to computing the maximal independent set and appear in the results.</span>
<span class="sd">    :return: A Table with the list of the samples to drop along with their rank.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Make sure that the key types are valid</span>
    <span class="k">assert</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span><span class="o">==</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">assert</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
    <span class="k">assert</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">rank_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>
    <span class="k">assert</span> <span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="n">rank_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Filtering related samples using a kin threshold of </span><span class="si">{kin_threshold}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">kin</span> <span class="o">&gt;</span> <span class="n">kin_threshold</span><span class="p">)</span>

    <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">min_related_hard_filter</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Computing samples related to too many individuals (&gt;</span><span class="si">{min_related_hard_filter}</span><span class="s2">) for exclusion&quot;</span><span class="p">)</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">)</span>
        <span class="n">gbi</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">())</span>
        <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="n">gbi</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">n</span> <span class="o">&gt;</span> <span class="n">min_related_hard_filter</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">gbi</span><span class="o">.</span><span class="n">s</span><span class="p">)))</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Found {len(filtered_samples_rel)} samples with too many 1st/2nd degree relatives. These samples will be excluded.&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">filtered_samples</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">filtered_samples_rel</span> <span class="o">=</span> <span class="n">filtered_samples_rel</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
            <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">collect_as_set</span><span class="p">(</span><span class="n">s</span><span class="p">),</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">))</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">s</span><span class="p">:</span> <span class="n">filtered_samples</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">s</span><span class="p">))</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">filtered_samples_lit</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span>
        <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">filtered_samples_lit</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">|</span>
            <span class="n">filtered_samples_lit</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">keep</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating related sample pairs with rank.&quot;</span><span class="p">)</span>
    <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">s</span><span class="p">,</span>
            <span class="n">rank</span><span class="o">=</span><span class="n">rank_ht</span><span class="p">[</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span>
        <span class="p">)}</span>
    <span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="n">j</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="n">s</span><span class="o">=</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">s</span><span class="p">,</span>
            <span class="n">rank</span><span class="o">=</span><span class="n">rank_ht</span><span class="p">[</span><span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">rank</span>
        <span class="p">)}</span>
    <span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">j</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;s&#39;</span><span class="p">)</span>
    <span class="n">relatedness_ht</span> <span class="o">=</span> <span class="n">relatedness_ht</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">maximal_independent_set</span><span class="p">(</span>
        <span class="n">relatedness_ht</span><span class="p">[</span><span class="n">i</span><span class="p">],</span>
        <span class="n">relatedness_ht</span><span class="p">[</span><span class="n">j</span><span class="p">],</span>
        <span class="n">keep</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="n">tie_breaker</span><span class="o">=</span><span class="k">lambda</span> <span class="n">l</span><span class="p">,</span><span class="n">r</span><span class="p">:</span> <span class="n">l</span><span class="o">.</span><span class="n">rank</span> <span class="o">-</span> <span class="n">r</span><span class="o">.</span><span class="n">rank</span>
    <span class="p">)</span>
    <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">()</span>
    <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">**</span><span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">node</span><span class="p">)</span>
    <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s1">&#39;s&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filtered_samples_rel</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">related_samples_to_drop_ht</span> <span class="o">=</span> <span class="n">related_samples_to_drop_ht</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="o">.</span><span class="n">parallelize</span><span class="p">(</span>
                <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="n">s</span><span class="o">=</span><span class="n">s</span><span class="p">,</span> <span class="n">rank</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">null</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tint64</span><span class="p">))</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">filtered_samples_rel</span><span class="p">],</span>
                <span class="n">key</span><span class="o">=</span><span class="s1">&#39;s&#39;</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">related_samples_to_drop_ht</span></div>


<div class="viewcode-block" id="run_pca_with_relateds"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.run_pca_with_relateds">[docs]</a><span class="k">def</span> <span class="nf">run_pca_with_relateds</span><span class="p">(</span>
        <span class="n">qc_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">related_samples_to_drop</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">],</span>
        <span class="n">n_pcs</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">10</span><span class="p">,</span>
        <span class="n">autosomes_only</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="nb">float</span><span class="p">],</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    First runs PCA excluding the given related samples,</span>
<span class="sd">    then projects these samples in the PC space to return scores for all samples.</span>

<span class="sd">    The `related_samples_to_drop` Table has to be keyed by the sample ID and all samples present in this</span>
<span class="sd">    table will be excluded from the PCA.</span>

<span class="sd">    The loadings Table returned also contains a `pca_af` annotation which is the allele frequency</span>
<span class="sd">    used for PCA. This is useful to project other samples in the PC space.</span>

<span class="sd">    :param qc_mt: Input QC MT</span>
<span class="sd">    :param related_samples_to_drop: Optional table of related samples to drop</span>
<span class="sd">    :param n_pcs: Number of PCs to compute</span>
<span class="sd">    :param autosomes_only: Whether to run the analysis on autosomes only</span>
<span class="sd">    :return: eigenvalues, scores and loadings</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">unrelated_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="k">if</span> <span class="n">autosomes_only</span><span class="p">:</span>
        <span class="n">unrelated_mt</span> <span class="o">=</span> <span class="n">filter_to_autosomes</span><span class="p">(</span><span class="n">unrelated_mt</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">related_samples_to_drop</span><span class="p">:</span>
        <span class="n">unrelated_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_missing</span><span class="p">(</span><span class="n">related_samples_to_drop</span><span class="p">[</span><span class="n">qc_mt</span><span class="o">.</span><span class="n">col_key</span><span class="p">]))</span>

    <span class="n">pca_evals</span><span class="p">,</span> <span class="n">pca_scores</span><span class="p">,</span> <span class="n">pca_loadings</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">hwe_normalized_pca</span><span class="p">(</span><span class="n">unrelated_mt</span><span class="o">.</span><span class="n">GT</span><span class="p">,</span> <span class="n">k</span><span class="o">=</span><span class="n">n_pcs</span><span class="p">,</span> <span class="n">compute_loadings</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">pca_af_ht</span> <span class="o">=</span> <span class="n">unrelated_mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">pca_af</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">unrelated_mt</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">n_alt_alleles</span><span class="p">())</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span>
    <span class="n">pca_loadings</span> <span class="o">=</span> <span class="n">pca_loadings</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">pca_af</span><span class="o">=</span><span class="n">pca_af_ht</span><span class="p">[</span><span class="n">pca_loadings</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">pca_af</span><span class="p">)</span>  <span class="c1"># TODO: Evaluate if needed to write results at this point if relateds or not</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">related_samples_to_drop</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">pca_evals</span><span class="p">,</span> <span class="n">pca_scores</span><span class="p">,</span> <span class="n">pca_loadings</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pca_loadings</span> <span class="o">=</span> <span class="n">pca_loadings</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="n">pca_scores</span> <span class="o">=</span> <span class="n">pca_scores</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="n">related_mt</span> <span class="o">=</span> <span class="n">qc_mt</span><span class="o">.</span><span class="n">filter_cols</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">related_samples_to_drop</span><span class="p">[</span><span class="n">qc_mt</span><span class="o">.</span><span class="n">col_key</span><span class="p">]))</span>
        <span class="n">related_scores</span> <span class="o">=</span> <span class="n">pc_project</span><span class="p">(</span><span class="n">related_mt</span><span class="p">,</span> <span class="n">pca_loadings</span><span class="p">)</span>
        <span class="n">pca_scores</span> <span class="o">=</span> <span class="n">pca_scores</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">related_scores</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">pca_evals</span><span class="p">,</span> <span class="n">pca_scores</span><span class="p">,</span> <span class="n">pca_loadings</span></div>


<div class="viewcode-block" id="compute_stratified_metrics_filter"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.compute_stratified_metrics_filter">[docs]</a><span class="k">def</span> <span class="nf">compute_stratified_metrics_filter</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">qc_metrics</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">],</span>
        <span class="n">strata</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Expression</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">lower_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">4.0</span><span class="p">,</span>
        <span class="n">upper_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">4.0</span><span class="p">,</span>
        <span class="n">metric_threshold</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">float</span><span class="p">,</span> <span class="nb">float</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">filter_name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;qc_metrics_filters&#39;</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Compute median, MAD, and upper and lower thresholds for each metric used in outlier filtering</span>

<span class="sd">    :param ht: HT containing relevant sample QC metric annotations</span>
<span class="sd">    :param qc_metrics: list of metrics (name and expr) for which to compute the critical values for filtering outliers</span>
<span class="sd">    :param strata: List of annotations used for stratification. These metrics should be discrete types!</span>
<span class="sd">    :param lower_threshold: Lower MAD threshold</span>
<span class="sd">    :param upper_threshold: Upper MAD threshold</span>
<span class="sd">    :param metric_threshold: Can be used to specify different (lower, upper) thresholds for one or more metrics</span>
<span class="sd">    :param filter_name: Name of resulting filters annotation</span>
<span class="sd">    :return: Table grouped by strata, with upper and lower threshold values computed for each sample QC metric</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_metric_threshold</span> <span class="o">=</span> <span class="p">{</span><span class="n">metric</span><span class="p">:</span> <span class="p">(</span><span class="n">lower_threshold</span><span class="p">,</span> <span class="n">upper_threshold</span><span class="p">)</span> <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">qc_metrics</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">metric_threshold</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">_metric_threshold</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">metric_threshold</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">make_filters_expr</span><span class="p">(</span><span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span> <span class="n">qc_metrics</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">x</span><span class="p">),</span>
                <span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;fail_</span><span class="si">{metric}</span><span class="s1">&#39;</span><span class="p">],</span> <span class="n">metric</span><span class="p">)</span> <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">qc_metrics</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">strata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">strata</span> <span class="o">=</span> <span class="p">{}</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="o">**</span><span class="n">qc_metrics</span><span class="p">,</span> <span class="o">**</span><span class="n">strata</span><span class="p">)</span><span class="o">.</span><span class="n">key_by</span><span class="p">(</span><span class="s1">&#39;s&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="n">agg_expr</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="o">**</span><span class="p">{</span>
        <span class="n">metric</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">bind</span><span class="p">(</span>
            <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                <span class="n">lower</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">median</span> <span class="o">-</span> <span class="n">_metric_threshold</span><span class="p">[</span><span class="n">metric</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">mad</span><span class="p">,</span>
                <span class="n">upper</span><span class="o">=</span><span class="n">x</span><span class="o">.</span><span class="n">median</span> <span class="o">+</span> <span class="n">_metric_threshold</span><span class="p">[</span><span class="n">metric</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">mad</span>
            <span class="p">),</span>
            <span class="n">get_median_and_mad_expr</span><span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="n">metric</span><span class="p">])</span>
        <span class="p">)</span> <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">qc_metrics</span>
    <span class="p">})</span>

    <span class="k">if</span> <span class="n">strata</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">qc_metrics_stats</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="n">ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strata</span><span class="p">]),</span> <span class="n">agg_expr</span><span class="p">),</span> <span class="n">_localize</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
        <span class="n">metrics_stats_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">qc_metrics_stats</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="n">ht</span><span class="p">[</span><span class="n">x</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strata</span><span class="p">])]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span><span class="n">qc_metrics_stats</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">agg_expr</span><span class="p">,</span> <span class="n">_localize</span><span class="o">=</span><span class="kc">False</span><span class="p">))</span>
        <span class="n">metrics_stats_expr</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">qc_metrics_stats</span>

    <span class="n">fail_exprs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="sa">f</span><span class="s1">&#39;fail_</span><span class="si">{metric}</span><span class="s1">&#39;</span><span class="p">:</span>
            <span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">metrics_stats_expr</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">lower</span><span class="p">)</span> <span class="o">|</span>
            <span class="p">(</span><span class="n">ht</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">metrics_stats_expr</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">upper</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">qc_metrics</span><span class="p">}</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span><span class="o">**</span><span class="n">fail_exprs</span><span class="p">)</span>
    <span class="n">stratified_filters</span> <span class="o">=</span> <span class="n">make_filters_expr</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">qc_metrics</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="o">**</span><span class="p">{</span><span class="n">filter_name</span><span class="p">:</span> <span class="n">stratified_filters</span><span class="p">})</span></div>


<div class="viewcode-block" id="compute_qc_metrics_residuals"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.compute_qc_metrics_residuals">[docs]</a><span class="k">def</span> <span class="nf">compute_qc_metrics_residuals</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">pc_scores</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">ArrayNumericExpression</span><span class="p">,</span>
        <span class="n">qc_metrics</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">],</span>
        <span class="n">use_pc_square</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">n_pcs</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">int</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">regression_sample_inclusion_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">bool</span><span class="p">(</span><span class="kc">True</span><span class="p">)</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes QC metrics residuals after regressing out PCs (and optionally PC^2)</span>

<span class="sd">    .. note::</span>

<span class="sd">        The `regression_sample_inclusion_expr` can be used to select a subset of the samples to include in the regression calculation.</span>
<span class="sd">        Residuals are always computed for all samples.</span>

<span class="sd">    :param ht: Input sample QC metrics HT</span>
<span class="sd">    :param pc_scores: The expression in the input HT that stores the PC scores</span>
<span class="sd">    :param qc_metrics: A dictionary with the name of each QC metric to compute residuals for and their corresponding expression in the input HT.</span>
<span class="sd">    :param use_pc_square: Whether to  use PC^2 in the regression or not</span>
<span class="sd">    :param n_pcs: Numer of PCs to use. If not set, then all PCs in `pc_scores` are used.</span>
<span class="sd">    :param regression_sample_inclusion_expr: An optional expression to select samples to include in the regression calculation.</span>
<span class="sd">    :return: Table with QC metrics residuals</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Annotate QC HT with fields necessary for computation</span>
    <span class="n">_sample_qc_ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="o">**</span><span class="n">qc_metrics</span><span class="p">,</span>
        <span class="n">scores</span><span class="o">=</span><span class="n">pc_scores</span><span class="p">,</span>
        <span class="n">_keep</span><span class="o">=</span><span class="n">regression_sample_inclusion_expr</span>
    <span class="p">)</span>

    <span class="c1"># If n_pcs wasn&#39;t provided, use all PCs</span>
    <span class="k">if</span> <span class="n">n_pcs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">n_pcs</span> <span class="o">=</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">len</span><span class="p">(</span><span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">)))</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Computing regressed QC metrics filters using </span><span class="si">{}</span><span class="s2"> PCs for metrics: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">n_pcs</span><span class="p">,</span>
        <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">qc_metrics</span><span class="p">)</span>
    <span class="p">))</span>

    <span class="c1"># Prepare regression variables, adding 1.0 first for the intercept</span>
    <span class="c1"># Adds square of variables if use_pc_square is true</span>
    <span class="n">x_expr</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_pcs</span><span class="p">)]</span>
    <span class="k">if</span> <span class="n">use_pc_square</span><span class="p">:</span>
        <span class="n">x_expr</span><span class="o">.</span><span class="n">extend</span><span class="p">([</span><span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">n_pcs</span><span class="p">)])</span>

    <span class="c1"># Compute linear regressions</span>
    <span class="n">lms</span> <span class="o">=</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span><span class="n">metric</span><span class="p">:</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">_keep</span><span class="p">,</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">linreg</span><span class="p">(</span>
                    <span class="n">y</span><span class="o">=</span><span class="n">_sample_qc_ht</span><span class="p">[</span><span class="n">metric</span><span class="p">],</span>
                    <span class="n">x</span><span class="o">=</span><span class="n">x_expr</span>
                <span class="p">)</span>
             <span class="p">)</span> <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">qc_metrics</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="n">_sample_qc_ht</span> <span class="o">=</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">lms</span><span class="o">=</span><span class="n">lms</span>
    <span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="c1"># Compute residuals</span>
    <span class="k">def</span> <span class="nf">get_lm_prediction_expr</span><span class="p">(</span><span class="n">metric</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">lm_pred_expr</span> <span class="o">=</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">lms</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">beta</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">n_pcs</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">lms</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">beta</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">use_pc_square</span><span class="p">:</span>
            <span class="n">lm_pred_expr</span> <span class="o">=</span> <span class="n">lm_pred_expr</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">range</span><span class="p">(</span><span class="n">n_pcs</span><span class="p">)</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">i</span><span class="p">:</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">lms</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">beta</span><span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="n">n_pcs</span> <span class="o">+</span> <span class="mi">1</span><span class="p">]</span> <span class="o">*</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">*</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">scores</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">lm_pred_expr</span>

    <span class="n">residuals_ht</span> <span class="o">=</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{metric}</span><span class="s1">_residual&#39;</span><span class="p">:</span> <span class="n">_sample_qc_ht</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span> <span class="o">-</span> <span class="n">get_lm_prediction_expr</span><span class="p">(</span><span class="n">metric</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">_sample_qc_ht</span><span class="o">.</span><span class="n">lms</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">residuals_ht</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span></div>


<div class="viewcode-block" id="add_filters_expr"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.add_filters_expr">[docs]</a><span class="k">def</span> <span class="nf">add_filters_expr</span><span class="p">(</span>
        <span class="n">filters</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">],</span>
        <span class="n">current_filters</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">SetExpression</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates an expression to create or add filters.</span>
<span class="sd">    For each entry in the `filters` dictionary, if the value evaluates to `True`,</span>
<span class="sd">    then the key is added as a filter name.</span>

<span class="sd">    Current filters are kept if provided using `current_filters`</span>

<span class="sd">    :param filters: The filters and their expressions</span>
<span class="sd">    :param current_filters: The set of current filters</span>
<span class="sd">    :return: An expression that can be used to annotate the filters</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">current_filters</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">current_filters</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">fold</span><span class="p">(</span>
        <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">y</span><span class="p">),</span>
        <span class="n">current_filters</span><span class="p">,</span>
        <span class="p">[</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">cond</span><span class="p">(</span><span class="n">filter_condition</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">set</span><span class="p">([</span><span class="n">filter_name</span><span class="p">]),</span> <span class="n">hl</span><span class="o">.</span><span class="n">empty_set</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tstr</span><span class="p">))</span>
            <span class="k">for</span> <span class="n">filter_name</span><span class="p">,</span> <span class="n">filter_condition</span> <span class="ow">in</span> <span class="n">filters</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">]</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="merge_sample_qc_expr"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.merge_sample_qc_expr">[docs]</a><span class="k">def</span> <span class="nf">merge_sample_qc_expr</span><span class="p">(</span><span class="n">sample_qc_exprs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">StructExpression</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Creates an expression that merges results from non-overlapping strata of hail.sample_qc</span>

<span class="sd">    E.g.:</span>

<span class="sd">    - Compute autosomes and sex chromosomes metrics separately, then merge results</span>
<span class="sd">    - Compute bi-allelic and multi-allelic metrics separately, then merge results</span>

<span class="sd">    Note regarding the merging of ``dp_stats`` and ``gq_stats``:</span>
<span class="sd">    Because ``n`` is needed to aggregate ``stdev``, ``n_called`` is used for this purpose.</span>
<span class="sd">    This should work very well on a standard GATK VCF and it essentially assumes that:</span>

<span class="sd">    - samples that are called have `DP` and `GQ` fields</span>
<span class="sd">    - samples that are not called do not have `DP` and `GQ` fields</span>

<span class="sd">    Even if these assumptions are broken for some genotypes, it shouldn&#39;t matter too much.</span>

<span class="sd">    :param sample_qc_exprs: List of sample QC struct expressions for each stratification</span>
<span class="sd">    :return: Combined sample QC results</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># List of metrics that can be aggregated by summing</span>
    <span class="n">additive_metrics</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s1">&#39;n_called&#39;</span><span class="p">,</span> <span class="s1">&#39;n_not_called&#39;</span><span class="p">,</span> <span class="s1">&#39;n_hom_ref&#39;</span><span class="p">,</span> <span class="s1">&#39;n_het&#39;</span><span class="p">,</span> <span class="s1">&#39;n_hom_var&#39;</span><span class="p">,</span> <span class="s1">&#39;n_snp&#39;</span><span class="p">,</span>
        <span class="s1">&#39;n_insertion&#39;</span><span class="p">,</span> <span class="s1">&#39;n_deletion&#39;</span><span class="p">,</span> <span class="s1">&#39;n_singleton&#39;</span><span class="p">,</span> <span class="s1">&#39;n_transition&#39;</span><span class="p">,</span> <span class="s1">&#39;n_transversion&#39;</span><span class="p">,</span> <span class="s1">&#39;n_star&#39;</span>
    <span class="p">]</span>

    <span class="c1"># List of metrics that are ratio of summed metrics (name, nominator, denominator)</span>
    <span class="n">ratio_metrics</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="s1">&#39;call_rate&#39;</span><span class="p">,</span> <span class="s1">&#39;n_called&#39;</span><span class="p">,</span> <span class="s1">&#39;n_not_called&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;r_ti_tv&#39;</span><span class="p">,</span> <span class="s1">&#39;n_transition&#39;</span><span class="p">,</span> <span class="s1">&#39;n_transversion&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;r_het_hom_var&#39;</span><span class="p">,</span> <span class="s1">&#39;n_het&#39;</span><span class="p">,</span> <span class="s1">&#39;n_hom_var&#39;</span><span class="p">),</span>
        <span class="p">(</span><span class="s1">&#39;r_insertion_deletion&#39;</span><span class="p">,</span> <span class="s1">&#39;n_insertion&#39;</span><span class="p">,</span> <span class="s1">&#39;n_deletion&#39;</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="c1"># List of metrics that are struct generated by a stats counter</span>
    <span class="n">stats_metrics</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;gq_stats&#39;</span><span class="p">,</span> <span class="s1">&#39;dp_stats&#39;</span><span class="p">]</span>

    <span class="c1"># Gather metrics present in sample qc fields</span>
    <span class="n">sample_qc_fields</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample_qc_exprs</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">sample_qc_expr</span> <span class="ow">in</span> <span class="n">sample_qc_exprs</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
        <span class="n">sample_qc_fields</span> <span class="o">=</span> <span class="n">sample_qc_fields</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">sample_qc_expr</span><span class="p">))</span>

    <span class="c1"># Merge additive metrics in sample qc fields</span>
    <span class="n">merged_exprs</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">metric</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">sum</span><span class="p">([</span><span class="n">sample_qc_expr</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span> <span class="k">for</span> <span class="n">sample_qc_expr</span> <span class="ow">in</span> <span class="n">sample_qc_exprs</span><span class="p">])</span>
        <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">additive_metrics</span> <span class="k">if</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">sample_qc_fields</span>
    <span class="p">}</span>

    <span class="c1"># Merge ratio metrics in sample qc fields</span>
    <span class="n">merged_exprs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span>
        <span class="n">metric</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">float64</span><span class="p">(</span><span class="n">divide_null</span><span class="p">(</span><span class="n">merged_exprs</span><span class="p">[</span><span class="n">nom</span><span class="p">],</span> <span class="n">merged_exprs</span><span class="p">[</span><span class="n">denom</span><span class="p">]))</span>
        <span class="k">for</span> <span class="n">metric</span><span class="p">,</span> <span class="n">nom</span><span class="p">,</span> <span class="n">denom</span> <span class="ow">in</span> <span class="n">ratio_metrics</span> <span class="k">if</span> <span class="n">nom</span> <span class="ow">in</span> <span class="n">sample_qc_fields</span> <span class="ow">and</span> <span class="n">denom</span> <span class="ow">in</span> <span class="n">sample_qc_fields</span>
    <span class="p">})</span>

    <span class="c1"># Merge stats counter metrics in sample qc fields</span>
    <span class="c1"># Use n_called as n for DP and GQ stats</span>
    <span class="k">if</span> <span class="s1">&#39;n_called&#39;</span> <span class="ow">in</span> <span class="n">sample_qc_fields</span><span class="p">:</span>
        <span class="n">merged_exprs</span><span class="o">.</span><span class="n">update</span><span class="p">({</span>
            <span class="n">metric</span><span class="p">:</span> <span class="n">merge_stats_counters_expr</span><span class="p">([</span>
                <span class="n">sample_qc_expr</span><span class="p">[</span><span class="n">metric</span><span class="p">]</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">n</span><span class="o">=</span><span class="n">sample_qc_expr</span><span class="o">.</span><span class="n">n_called</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">sample_qc_expr</span> <span class="ow">in</span> <span class="n">sample_qc_exprs</span>
            <span class="p">])</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;n&#39;</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">metric</span> <span class="ow">in</span> <span class="n">stats_metrics</span>
        <span class="p">})</span>

    <span class="k">return</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span><span class="o">**</span><span class="n">merged_exprs</span><span class="p">)</span></div>


<div class="viewcode-block" id="compute_stratified_sample_qc"><a class="viewcode-back" href="../../../api_reference/utils/sample_qc.html#gnomad.utils.sample_qc.compute_stratified_sample_qc">[docs]</a><span class="k">def</span> <span class="nf">compute_stratified_sample_qc</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">strata</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">],</span>
        <span class="n">tmp_ht_prefix</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
        <span class="n">gt_expr</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">CallExpression</span><span class="p">]</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Runs hl.sample_qc on different strata and then also merge the results into a single expression.</span>
<span class="sd">    Note that strata should be non-overlapping,  e.g. SNV vs indels or bi-allelic vs multi-allelic</span>

<span class="sd">    :param mt: Input MT</span>
<span class="sd">    :param strata: Strata names and filtering expressions</span>
<span class="sd">    :param tmp_ht_prefix: Optional path prefix to write the intermediate strata results to (recommended for larger datasets)</span>
<span class="sd">    :param gt_expr: Optional entry field storing the genotype (if not specified, then it is assumed that it is stored in mt.GT)</span>
<span class="sd">    :return: Sample QC table, including strat-specific numbers</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">select_rows</span><span class="p">(</span>
        <span class="o">**</span><span class="n">strata</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">gt_expr</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="n">GT</span><span class="o">=</span><span class="n">gt_expr</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">mt</span> <span class="o">=</span> <span class="n">mt</span><span class="o">.</span><span class="n">select_entries</span><span class="p">(</span><span class="s1">&#39;GT&#39;</span><span class="p">)</span>

    <span class="n">strat_hts</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">strat</span> <span class="ow">in</span> <span class="n">strata</span><span class="p">:</span>
        <span class="n">strat_sample_qc_ht</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">sample_qc</span><span class="p">(</span><span class="n">mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span><span class="n">mt</span><span class="p">[</span><span class="n">strat</span><span class="p">]))</span><span class="o">.</span><span class="n">cols</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">tmp_ht_prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">strat_sample_qc_ht</span> <span class="o">=</span> <span class="n">strat_sample_qc_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">tmp_ht_prefix</span> <span class="o">+</span> <span class="sa">f</span><span class="s1">&#39;_</span><span class="si">{strat}</span><span class="s1">.ht&#39;</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">strat_sample_qc_ht</span> <span class="o">=</span> <span class="n">strat_sample_qc_ht</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="n">strat_hts</span><span class="p">[</span><span class="n">strat</span><span class="p">]</span> <span class="o">=</span> <span class="n">strat_sample_qc_ht</span>

    <span class="n">sample_qc_ht</span> <span class="o">=</span> <span class="n">strat_hts</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">strata</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">sample_qc_ht</span> <span class="o">=</span> <span class="n">sample_qc_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;{list(strata)[0]}_sample_qc&#39;</span><span class="p">:</span> <span class="n">sample_qc_ht</span><span class="o">.</span><span class="n">sample_qc</span><span class="p">},</span>
        <span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{strat}</span><span class="s1">_sample_qc&#39;</span><span class="p">:</span> <span class="n">strat_hts</span><span class="p">[</span><span class="n">strat</span><span class="p">][</span><span class="n">sample_qc_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">sample_qc</span> <span class="k">for</span> <span class="n">strat</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">strata</span><span class="p">)[</span><span class="mi">1</span><span class="p">:]}</span>
    <span class="p">)</span>
    <span class="n">sample_qc_ht</span> <span class="o">=</span> <span class="n">sample_qc_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">sample_qc</span><span class="o">=</span><span class="n">merge_sample_qc_expr</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">sample_qc_ht</span><span class="o">.</span><span class="n">row_value</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">sample_qc_ht</span></div>

</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>