

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>gnomad.utils.variant_qc &mdash; gnomad master documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/theme_overrides.css" type="text/css" />
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> gnomad
          

          
          </a>

          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api_reference/index.html">API Reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">gnomad</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>gnomad.utils.variant_qc</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for gnomad.utils.variant_qc</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">.generic</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">gnomad.resources.grch37</span> <span class="k">as</span> <span class="nn">grch37_resources</span>
<span class="kn">import</span> <span class="nn">gnomad.resources.grch38</span> <span class="k">as</span> <span class="nn">grch38_resources</span>


<div class="viewcode-block" id="compute_quantile_bin"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.compute_quantile_bin">[docs]</a><span class="k">def</span> <span class="nf">compute_quantile_bin</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">score_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">NumericExpression</span><span class="p">,</span>
        <span class="n">bin_expr</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;bin&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
        <span class="n">compute_snv_indel_separately</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">n_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
        <span class="n">k</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1000</span><span class="p">,</span>
        <span class="n">desc</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Returns a table with a bin for each row based on quantiles of `score_expr`.</span>

<span class="sd">    The bin is computed by dividing the `score_expr` into `n_bins` bins containing an equal number of elements.</span>
<span class="sd">    This is done based on quantiles computed with hl.agg.approx_quantiles. If a single value in `score_expr` spans more</span>
<span class="sd">    than one bin, the rows with this value are distributed randomly across the bins it spans.</span>

<span class="sd">    If `compute_snv_indel_separately` is True all items in `bin_expr` will be stratified by snv / indels for the bin</span>
<span class="sd">    calculation. Because SNV and indel rows are mutually exclusive, they are re-combined into a single annotation. For</span>
<span class="sd">    example if we have the following four variants and scores and `n_bins` of 2:</span>

<span class="sd">    ========   =======   ======   =================   =================</span>
<span class="sd">    Variant    Type      Score    bin - `compute_snv_indel_separately`:</span>
<span class="sd">    --------   -------   ------   -------------------------------------</span>
<span class="sd">    \          \         \        False               True</span>
<span class="sd">    ========   =======   ======   =================   =================</span>
<span class="sd">    Var1       SNV       0.1      1                   1</span>
<span class="sd">    Var2       SNV       0.2      1                   2</span>
<span class="sd">    Var3       Indel     0.3      2                   1</span>
<span class="sd">    Var4       Indel     0.4      2                   2</span>
<span class="sd">    ========   =======   ======   =================   =================</span>

<span class="sd">    .. note::</span>

<span class="sd">        The `bin_expr` defines which data the bin(s) should be computed on. E.g., to get a biallelic quantile bin and an</span>
<span class="sd">        singleton quantile bin, the following could be used:</span>

<span class="sd">        .. code-block:: python</span>

<span class="sd">            bin_expr={</span>
<span class="sd">                &#39;biallelic_bin&#39;: ~ht.was_split,</span>
<span class="sd">                &#39;singleton_bin&#39;: ht.singleton</span>
<span class="sd">            }</span>

<span class="sd">    :param ht: Input Table</span>
<span class="sd">    :param score_expr: Expression containing the score</span>
<span class="sd">    :param bin_expr: Quantile bin(s) to be computed (see notes)</span>
<span class="sd">    :param compute_snv_indel_separately: Should all `bin_expr` items be stratified by snv / indels</span>
<span class="sd">    :param n_bins: Number of bins to bin the data into</span>
<span class="sd">    :param k: The `k` parameter of approx_quantiles</span>
<span class="sd">    :param desc: Whether to bin the score in descending order</span>
<span class="sd">    :return: Table with the quantile bins</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">import</span> <span class="nn">math</span>

    <span class="k">def</span> <span class="nf">quantiles_to_bin_boundaries</span><span class="p">(</span><span class="n">quantiles</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">int</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Merges bins with the same boundaries into a unique bin while keeping track of</span>
<span class="sd">        which bins have been merged and the global index of all bins.</span>

<span class="sd">        :param quantiles: Original bins boundaries</span>
<span class="sd">        :return: (dict of the indices of bins for which multiple bins were collapsed -&gt; number of bins collapsed,</span>
<span class="sd">                  Global indices of merged bins,</span>
<span class="sd">                  Merged bins boundaries)</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Pad the quantiles to create boundaries for the first and last bins</span>
        <span class="n">bin_boundaries</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="n">math</span><span class="o">.</span><span class="n">inf</span><span class="p">]</span> <span class="o">+</span> <span class="n">quantiles</span> <span class="o">+</span> <span class="p">[</span><span class="n">math</span><span class="o">.</span><span class="n">inf</span><span class="p">]</span>
        <span class="n">merged_bins</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="c1"># If every quantile has a unique value, then bin boudaries are unique</span>
        <span class="c1"># and can be passed to binary_search as-is</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">quantiles</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">quantiles</span><span class="p">)):</span>
            <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
                <span class="n">merged_bins</span><span class="o">=</span><span class="n">merged_bins</span><span class="p">,</span>
                <span class="n">global_bin_indices</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">bin_boundaries</span><span class="p">))),</span>
                <span class="n">bin_boundaries</span><span class="o">=</span><span class="n">bin_boundaries</span>
            <span class="p">)</span>

        <span class="n">indexed_bins</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">enumerate</span><span class="p">(</span><span class="n">bin_boundaries</span><span class="p">))</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">indexed_bins</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">indexed_bins</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">indexed_bins</span><span class="p">[</span><span class="n">i</span><span class="p">][</span><span class="mi">1</span><span class="p">]:</span>
                <span class="n">merged_bins</span><span class="p">[</span><span class="n">i</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">indexed_bins</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>

        <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
            <span class="n">merged_bins</span><span class="o">=</span><span class="n">merged_bins</span><span class="p">,</span>
            <span class="n">global_bin_indices</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">indexed_bins</span><span class="p">],</span>
            <span class="n">bin_boundaries</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">indexed_bins</span><span class="p">]</span>
        <span class="p">)</span>

    <span class="k">if</span> <span class="n">compute_snv_indel_separately</span><span class="p">:</span>
        <span class="c1"># For each bin, add a SNV / indel stratification</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="p">{</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{bin_id}</span><span class="s1">_</span><span class="si">{snv}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="p">(</span><span class="n">bin_expr</span> <span class="o">&amp;</span> <span class="n">snv_expr</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">bin_id</span><span class="p">,</span> <span class="n">bin_expr</span> <span class="ow">in</span> <span class="n">bin_expr</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">snv</span><span class="p">,</span> <span class="n">snv_expr</span> <span class="ow">in</span> <span class="p">[</span>
                <span class="p">(</span><span class="s1">&#39;snv&#39;</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">snv</span><span class="p">),</span>
                <span class="p">(</span><span class="s1">&#39;indel&#39;</span><span class="p">,</span> <span class="o">~</span><span class="n">ht</span><span class="o">.</span><span class="n">snv</span><span class="p">)</span>
            <span class="p">]</span>
        <span class="p">}</span>

    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span><span class="sa">f</span><span class="s1">&#39;_filter_</span><span class="si">{bin_id}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">bin_expr</span> <span class="k">for</span> <span class="n">bin_id</span><span class="p">,</span> <span class="n">bin_expr</span> <span class="ow">in</span> <span class="n">bin_expr</span><span class="o">.</span><span class="n">items</span><span class="p">()},</span>
        <span class="n">_score</span><span class="o">=</span><span class="n">score_expr</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Adding quantile bins using approximate_quantiles binned into </span><span class="si">{n_bins}</span><span class="s1">, using k=</span><span class="si">{k}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">bin_stats</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                    <span class="n">bin_ht</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;_filter_</span><span class="si">{bin_id}</span><span class="s1">&#39;</span><span class="p">],</span>
                    <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="n">n</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
                        <span class="n">quantiles</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">approx_quantiles</span><span class="p">(</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">_score</span><span class="p">,</span> <span class="p">[</span><span class="n">x</span> <span class="o">/</span> <span class="p">(</span><span class="n">n_bins</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_bins</span><span class="p">)],</span> <span class="n">k</span><span class="o">=</span><span class="n">k</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr</span>
            <span class="p">}</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Take care of bins with duplicated boundaries</span>
    <span class="n">bin_stats</span> <span class="o">=</span> <span class="n">bin_stats</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="n">rname</span><span class="p">:</span> <span class="n">bin_stats</span><span class="p">[</span><span class="n">rname</span><span class="p">]</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
                    <span class="o">**</span><span class="n">quantiles_to_bin_boundaries</span><span class="p">(</span><span class="n">bin_stats</span><span class="p">[</span><span class="n">rname</span><span class="p">]</span><span class="o">.</span><span class="n">quantiles</span><span class="p">)</span>
            <span class="p">)</span> <span class="k">for</span> <span class="n">rname</span> <span class="ow">in</span> <span class="n">bin_stats</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
        <span class="n">bin_stats</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">literal</span><span class="p">(</span>
            <span class="n">bin_stats</span><span class="p">,</span>
            <span class="n">dtype</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tstruct</span><span class="p">(</span><span class="o">**</span><span class="p">{</span>
                <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">tstruct</span><span class="p">(</span>
                    <span class="n">n</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tint64</span><span class="p">,</span>
                    <span class="n">quantiles</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tarray</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tfloat64</span><span class="p">),</span>
                    <span class="n">bin_boundaries</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tarray</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tfloat64</span><span class="p">),</span>
                    <span class="n">global_bin_indices</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tarray</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tint32</span><span class="p">),</span>
                    <span class="n">merged_bins</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">tdict</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">tint32</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">tint32</span><span class="p">)</span>
                <span class="p">)</span> <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr</span>
            <span class="p">})</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="c1"># Annotate the bin as the index in the unique boundaries array</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_missing</span><span class="p">(</span>
                <span class="n">bin_ht</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;_filter_</span><span class="si">{bin_id}</span><span class="s1">&#39;</span><span class="p">],</span>
                <span class="n">hl</span><span class="o">.</span><span class="n">binary_search</span><span class="p">(</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span><span class="o">.</span><span class="n">bin_boundaries</span><span class="p">,</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">_score</span><span class="p">),</span>
            <span class="p">)</span> <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="c1"># Convert the bin to global bin by expanding merged bins, that is:</span>
    <span class="c1"># If a value falls in a bin that needs expansion, assign it randomly to one of the expanded bins</span>
    <span class="c1"># Otherwise, simply modify the bin to its global index (with expanded bins that is)</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
        <span class="o">**</span><span class="p">{</span>
            <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">cond</span><span class="p">(</span>
                <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span><span class="o">.</span><span class="n">merged_bins</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]),</span>
                <span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span> <span class="o">+</span> <span class="n">hl</span><span class="o">.</span><span class="n">int</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">rand_unif</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span><span class="o">.</span><span class="n">merged_bins</span><span class="p">[</span><span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]]</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)),</span>
                <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span><span class="o">.</span><span class="n">global_bin_indices</span><span class="p">[</span><span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">desc</span><span class="p">:</span>
        <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span><span class="n">bin_id</span><span class="p">:</span> <span class="n">n_bins</span> <span class="o">-</span> <span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_id</span><span class="p">]</span> <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr</span><span class="p">}</span>
        <span class="p">)</span>

    <span class="c1"># Because SNV and indel rows are mutually exclusive, re-combine them into a single bin.</span>
    <span class="c1"># Update the global bin_stats struct to reflect the change in bin names in the table</span>
    <span class="k">if</span> <span class="n">compute_snv_indel_separately</span><span class="p">:</span>
        <span class="n">bin_expr_no_snv</span> <span class="o">=</span> <span class="p">{</span><span class="n">bin_id</span><span class="o">.</span><span class="n">rsplit</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">}</span>
        <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">annotate_globals</span><span class="p">(</span>
            <span class="n">bin_stats</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="o">**</span><span class="p">{</span>
                    <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                        <span class="o">**</span><span class="p">{</span><span class="n">snv</span><span class="p">:</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{bin_id}</span><span class="s2">_</span><span class="si">{snv}</span><span class="s2">&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">snv</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;snv&quot;</span><span class="p">,</span> <span class="s2">&quot;indel&quot;</span><span class="p">]}</span>
                    <span class="p">)</span>
                    <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr_no_snv</span>
                <span class="p">}</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
            <span class="o">**</span><span class="p">{</span>
                <span class="n">bin_id</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">cond</span><span class="p">(</span>
                    <span class="n">ht</span><span class="p">[</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">snv</span><span class="p">,</span>
                    <span class="n">bin_ht</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{bin_id}</span><span class="s1">_snv&#39;</span><span class="p">],</span>
                    <span class="n">bin_ht</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{bin_id}</span><span class="s1">_indel&#39;</span><span class="p">]</span>
                <span class="p">)</span>
                <span class="k">for</span> <span class="n">bin_id</span> <span class="ow">in</span> <span class="n">bin_expr_no_snv</span>
            <span class="p">}</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">bin_ht</span></div>


<div class="viewcode-block" id="default_create_binned_ht"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.default_create_binned_ht">[docs]</a><span class="k">def</span> <span class="nf">default_create_binned_ht</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">n_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span><span class="p">,</span>
        <span class="n">singleton</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">biallelic</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">adj</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">add_substrat</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is meant as a default wrapper for `compute_quantile_bin`. It annotates table with a bin, where variants are</span>
<span class="sd">    binned based on score into `n_bins` equally-sized bins.</span>

<span class="sd">    .. note::</span>

<span class="sd">        The following fields should be present:</span>
<span class="sd">            - score</span>
<span class="sd">            - ac - expected that this is the adj filtered allele count</span>
<span class="sd">            - ac_raw - expected that this is the raw allele count before adj filtering</span>

<span class="sd">    Computes bin numbers stratified by SNV / Indels and with the following optional sub bins</span>
<span class="sd">        - singletons</span>
<span class="sd">        - biallelics</span>
<span class="sd">        - biallelic singletons</span>
<span class="sd">        - adj</span>
<span class="sd">        - adj biallelics</span>
<span class="sd">        - adj singletons</span>
<span class="sd">        - adj biallelic singletons</span>

<span class="sd">    :param ht: Input table</span>
<span class="sd">    :param n_bins: Number of bins to bin into</span>
<span class="sd">    :param singleton: Should bins be stratified by singletons</span>
<span class="sd">    :param biallelic: Should bins be stratified by bi-alleleic variants</span>
<span class="sd">    :param adj: Should bins be stratified by adj filtering</span>
<span class="sd">    :param add_substrat: Any additional stratifications for adding bins</span>
<span class="sd">    :return: table with bin number for each variant</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">update_bin_expr</span><span class="p">(</span>
            <span class="n">bin_expr</span><span class="p">:</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">],</span>
            <span class="n">new_expr</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">,</span>
            <span class="n">new_id</span><span class="p">:</span> <span class="nb">str</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">BooleanExpression</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Updates a dictionary of expressions to add another stratification</span>

<span class="sd">        :param bin_expr: Dictionary of expressions to add another</span>
<span class="sd">        stratification to</span>
<span class="sd">        :param new_expr: New Boolean expression to add to `bin_expr`</span>
<span class="sd">        :param new_id: Name to add to each current key in `bin_expr` to indicate the new stratification</span>
<span class="sd">        :return: Dictionary of `bin_expr` updated with `new_expr` added as an additional stratification to all</span>
<span class="sd">        expressions already in `bin_expr`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">bin_expr</span><span class="o">.</span><span class="n">update</span><span class="p">({</span>
            <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{new_id}</span><span class="s1">_</span><span class="si">{bin_id}</span><span class="s1">&#39;</span><span class="p">:</span> <span class="n">bin_expr</span> <span class="o">&amp;</span> <span class="n">new_expr</span>
            <span class="k">for</span> <span class="n">bin_id</span><span class="p">,</span> <span class="n">bin_expr</span> <span class="ow">in</span> <span class="n">bin_expr</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
        <span class="p">})</span>
        <span class="k">return</span> <span class="n">bin_expr</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">singleton</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">ac_raw</span> <span class="o">==</span> <span class="mi">1</span><span class="p">,</span>
        <span class="n">snv</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>

    <span class="p">)</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
        <span class="n">ht</span><span class="o">.</span><span class="n">ac_raw</span> <span class="o">&gt;</span> <span class="mi">0</span>
    <span class="p">)</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="c1"># Desired bins and sub-bins</span>
    <span class="n">bin_expr</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;bin&#39;</span><span class="p">:</span> <span class="kc">True</span>
    <span class="p">}</span>

    <span class="k">if</span> <span class="n">singleton</span><span class="p">:</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">update_bin_expr</span><span class="p">(</span><span class="n">bin_expr</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">singleton</span><span class="p">,</span> <span class="s2">&quot;singleton&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">biallelic</span><span class="p">:</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">update_bin_expr</span><span class="p">(</span><span class="n">bin_expr</span><span class="p">,</span> <span class="o">~</span><span class="n">ht</span><span class="o">.</span><span class="n">was_split</span><span class="p">,</span> <span class="s2">&quot;biallelic&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">adj</span><span class="p">:</span>
        <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">update_bin_expr</span><span class="p">(</span><span class="n">bin_expr</span><span class="p">,</span> <span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ac</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span> <span class="s2">&quot;adj&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">add_substrat</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">add_id</span><span class="p">,</span> <span class="n">add_expr</span> <span class="ow">in</span> <span class="n">add_substrat</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">bin_expr</span> <span class="o">=</span> <span class="n">update_bin_expr</span><span class="p">(</span><span class="n">bin_expr</span><span class="p">,</span> <span class="n">add_expr</span><span class="p">,</span> <span class="n">add_id</span><span class="p">)</span>

    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">compute_quantile_bin</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">,</span>
        <span class="n">score_expr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
        <span class="n">bin_expr</span><span class="o">=</span><span class="n">bin_expr</span><span class="p">,</span>
        <span class="n">n_bins</span><span class="o">=</span><span class="n">n_bins</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bin_ht</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>


<div class="viewcode-block" id="default_score_bin_agg"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.default_score_bin_agg">[docs]</a><span class="k">def</span> <span class="nf">default_score_bin_agg</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">GroupedTable</span><span class="p">,</span>
        <span class="n">fam_stats_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">Aggregation</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Default aggregation function to add aggregations for min/max of score, number of ClinVar variants, number of truth</span>
<span class="sd">    variants (omni, mills, hapmap, and kgp_phase1), and family statistics.</span>

<span class="sd">    .. note::</span>

<span class="sd">        This function uses `ht._parent` to get the origin Table from the GroupedTable for the aggregation</span>

<span class="sd">    This can easily be combined with the GroupedTable returned by `compute_grouped_binned_ht`</span>

<span class="sd">    Example:</span>

<span class="sd">    .. code-block:: python</span>

<span class="sd">        binned_ht = create_binned_ht(...)</span>
<span class="sd">        grouped_binned_ht = compute_grouped_binned_ht(binned_ht)</span>
<span class="sd">        agg_ht = grouped_binned_ht.aggregate(default_score_bin_agg(**grouped_binned_ht, ...))</span>

<span class="sd">    .. note::</span>

<span class="sd">        The following annotations should be present:</span>

<span class="sd">        In ht:</span>
<span class="sd">            - score</span>
<span class="sd">            - singleton</span>
<span class="sd">            - positive_train_site</span>
<span class="sd">            - negative_train_site</span>
<span class="sd">            - ac_raw - expected that this is the raw allele count before adj filtering</span>
<span class="sd">            - info - struct that includes QD, FS, and MQ in order to add an annotation for fail_hard_filters</span>

<span class="sd">        In truth_ht:</span>
<span class="sd">            - omni</span>
<span class="sd">            - mills</span>
<span class="sd">            - hapmap</span>
<span class="sd">            - kgp_phase1_hc</span>

<span class="sd">        In fam_stats_ht:</span>
<span class="sd">            - n_de_novos_adj</span>
<span class="sd">            - n_de_novos_raw</span>
<span class="sd">            - n_transmitted_raw</span>
<span class="sd">            - unrelated_qc_callstats</span>
<span class="sd">            - tdt</span>

<span class="sd">    Automatic aggregations that will be done are:</span>
<span class="sd">        - `min_score` - minimun of score annotation per group</span>
<span class="sd">        - `max_score` - maiximum of score annotation per group</span>
<span class="sd">        - `n` - count of variants per group</span>
<span class="sd">        - `n_ins` - count of insertion per group</span>
<span class="sd">        - `n_ins` - count of insertion per group</span>
<span class="sd">        - `n_del` - count of deletions per group</span>
<span class="sd">        - `n_ti` - count of transitions per group</span>
<span class="sd">        - `n_tv` - count of trnasversions per group</span>
<span class="sd">        - `n_1bp_indel` - count of one base pair indels per group</span>
<span class="sd">        - `n_mod3bp_indel` - count of indels with a length divisible by three per group</span>
<span class="sd">        - `n_singleton` - count of singletons per group</span>
<span class="sd">        - `fail_hard_filters` - count of variants per group with QD &lt; 2 | FS &gt; 60 | MQ &lt; 30</span>
<span class="sd">        - `n_vqsr_pos_train` - count of variants that were a VQSR positive train site per group</span>
<span class="sd">        - `n_vqsr_neg_train` - count of variants that were a VQSR negative train site per group</span>
<span class="sd">        - `n_clinvar` - count of clinvar variants</span>
<span class="sd">        - `n_de_novos_adj` - count of adj filtered de dovo variants</span>
<span class="sd">        - `n_de_novos` - count of raw unfilterd filtered de dovo variants</span>
<span class="sd">        - `n_trans_singletons` - count of transmitted singletons</span>
<span class="sd">        - `n_untrans_singletons` - count of untransmitted singletons</span>
<span class="sd">        - `n_omni` - count of omni truth variants</span>
<span class="sd">        - `n_mills` - count of mills truth variants</span>
<span class="sd">        - `n_hapmap` - count of hapmap truth variants</span>
<span class="sd">        - `n_kgp_phase1_hc` - count of 1000 genomes phase 1 high confidence truth variants</span>

<span class="sd">    :param ht: Table that aggregation will be performed on</span>
<span class="sd">    :param fam_stats_ht: Path to family statistics HT</span>
<span class="sd">    :return: a dictionary containing aggregations to perform on ht</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Annotate binned table with the evaluation data</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">_parent</span>
    <span class="n">indel_length</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">abs</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">length</span><span class="p">()</span> <span class="o">-</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">length</span><span class="p">())</span>
    <span class="c1"># Load external evaluation data</span>
    <span class="n">build</span> <span class="o">=</span> <span class="n">get_reference_genome</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">locus</span><span class="p">)</span><span class="o">.</span><span class="n">name</span>
    <span class="n">clinvar</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">grch37_resources</span><span class="o">.</span><span class="n">reference_data</span><span class="o">.</span><span class="n">clinvar</span>
        <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;GRCh37&quot;</span>
        <span class="k">else</span> <span class="n">grch38_resources</span><span class="o">.</span><span class="n">reference_data</span><span class="o">.</span><span class="n">clinvar</span>
    <span class="p">)</span><span class="o">.</span><span class="n">ht</span><span class="p">()[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="n">truth_data</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">grch37_resources</span><span class="o">.</span><span class="n">reference_data</span><span class="o">.</span><span class="n">get_truth_ht</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">build</span> <span class="o">==</span> <span class="s2">&quot;GRCh37&quot;</span>
        <span class="k">else</span> <span class="n">grch38_resources</span><span class="o">.</span><span class="n">reference_data</span><span class="o">.</span><span class="n">get_truth_ht</span><span class="p">()</span>
    <span class="p">)[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="n">fam</span> <span class="o">=</span> <span class="n">fam_stats_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>

    <span class="k">return</span> <span class="nb">dict</span><span class="p">(</span>
        <span class="n">min_score</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">),</span>
        <span class="n">max_score</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">),</span>
        <span class="n">n</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">(),</span>
        <span class="n">n_ins</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_insertion</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])),</span>
        <span class="n">n_del</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_deletion</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])),</span>
        <span class="n">n_ti</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_transition</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])),</span>
        <span class="n">n_tv</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_transversion</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])),</span>
        <span class="n">n_1bp_indel</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">indel_length</span> <span class="o">==</span> <span class="mi">1</span><span class="p">),</span>
        <span class="n">n_mod3bp_indel</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">((</span><span class="n">indel_length</span> <span class="o">%</span> <span class="mi">3</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">),</span>
        <span class="n">n_singleton</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">singleton</span><span class="p">),</span>
        <span class="n">fail_hard_filters</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">((</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">QD</span> <span class="o">&lt;</span> <span class="mi">2</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">FS</span> <span class="o">&gt;</span> <span class="mi">60</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">info</span><span class="o">.</span><span class="n">MQ</span> <span class="o">&lt;</span> <span class="mi">30</span><span class="p">)),</span>
        <span class="n">n_vqsr_pos_train</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">positive_train_site</span><span class="p">),</span>
        <span class="n">n_vqsr_neg_train</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">negative_train_site</span><span class="p">),</span>
        <span class="n">n_clinvar</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">clinvar</span><span class="p">)),</span>
        <span class="n">n_de_novos_adj</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">fam</span><span class="o">.</span><span class="n">n_de_novos_adj</span><span class="p">),</span>
        <span class="n">n_de_novo</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">fam</span><span class="o">.</span><span class="n">n_de_novos_raw</span><span class="p">),</span>
        <span class="n">n_trans_singletons</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">ac_raw</span> <span class="o">==</span> <span class="mi">2</span><span class="p">,</span> <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">fam</span><span class="o">.</span><span class="n">n_transmitted_raw</span><span class="p">)),</span>
        <span class="n">n_untrans_singletons</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">filter</span><span class="p">((</span><span class="n">ht</span><span class="o">.</span><span class="n">ac_raw</span> <span class="o">&lt;</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">fam</span><span class="o">.</span><span class="n">unrelated_qc_callstats</span><span class="o">.</span><span class="n">AC</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="p">),</span>
                                           <span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">fam</span><span class="o">.</span><span class="n">tdt</span><span class="o">.</span><span class="n">u</span><span class="p">)),</span>
        <span class="c1"># n_train_trans_singletons=hl.agg.filter((ht.ac_raw == 2) &amp; rank_ht.positive_train_site, hl.agg.sum(fam.n_transmitted_raw)),</span>
        <span class="n">n_omni</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">truth_data</span><span class="o">.</span><span class="n">omni</span><span class="p">),</span>
        <span class="n">n_mills</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">truth_data</span><span class="o">.</span><span class="n">mills</span><span class="p">),</span>
        <span class="n">n_hapmap</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">truth_data</span><span class="o">.</span><span class="n">hapmap</span><span class="p">),</span>
        <span class="n">n_kgp_phase1_hc</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">truth_data</span><span class="o">.</span><span class="n">kgp_phase1_hc</span><span class="p">)</span>
    <span class="p">)</span></div>


<div class="viewcode-block" id="compute_grouped_binned_ht"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.compute_grouped_binned_ht">[docs]</a><span class="k">def</span> <span class="nf">compute_grouped_binned_ht</span><span class="p">(</span>
        <span class="n">bin_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">checkpoint_path</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">GroupedTable</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Groups a Table that has been annotated with bins based on quantiles (`compute_quantile_bin` or</span>
<span class="sd">    `default_create_binned_ht`). The table will be grouped by bin_id (bin, biallelic, etc.), contig, snv, bi_allelic and</span>
<span class="sd">    singleton.</span>

<span class="sd">    .. note::</span>

<span class="sd">        If performing an aggregation following this grouping (such as `default_score_bin_agg`) then the aggregation</span>
<span class="sd">        function will need to use `ht._parent` to get the origin Table from the GroupedTable for the aggregation</span>

<span class="sd">    :param bin_ht: Input Table with a `bin_id` annotation</span>
<span class="sd">    :param checkpoint_path: If provided an intermediate checkpoint table is created with all required annotations before shuffling.</span>
<span class="sd">    :return: Table grouped by bins(s)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Explode the rank table by bin_id</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">quantile_bins</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">array</span><span class="p">([</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">Struct</span><span class="p">(</span>
                <span class="n">bin_id</span><span class="o">=</span><span class="n">bin_name</span><span class="p">,</span>
                <span class="nb">bin</span><span class="o">=</span><span class="n">bin_ht</span><span class="p">[</span><span class="n">bin_name</span><span class="p">]</span>
            <span class="p">)</span>
            <span class="k">for</span> <span class="n">bin_name</span> <span class="ow">in</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_stats</span>
        <span class="p">])</span>
    <span class="p">)</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">quantile_bins</span><span class="p">)</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">transmute</span><span class="p">(</span>
        <span class="n">bin_id</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">quantile_bins</span><span class="o">.</span><span class="n">bin_id</span><span class="p">,</span>
        <span class="nb">bin</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">quantile_bins</span><span class="o">.</span><span class="n">bin</span>
    <span class="p">)</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">bin</span><span class="p">))</span>

    <span class="k">if</span> <span class="n">checkpoint_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">bin_ht</span><span class="o">.</span><span class="n">checkpoint</span><span class="p">(</span><span class="n">checkpoint_path</span><span class="p">,</span> <span class="n">overwrite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>

    <span class="c1"># Group by bin_id, bin and additional stratification desired and compute QC metrics per bin</span>
    <span class="k">return</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span>
            <span class="n">bin_id</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">bin_id</span><span class="p">,</span>
            <span class="n">contig</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">locus</span><span class="o">.</span><span class="n">contig</span><span class="p">,</span>
            <span class="n">snv</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">bin_ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">]),</span>
            <span class="n">bi_allelic</span><span class="o">=~</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">was_split</span><span class="p">,</span>
            <span class="n">singleton</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">singleton</span><span class="p">,</span>
            <span class="n">release_adj</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">ac</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span>
            <span class="nb">bin</span><span class="o">=</span><span class="n">bin_ht</span><span class="o">.</span><span class="n">bin</span>
        <span class="p">)</span><span class="o">.</span><span class="n">_set_buffer_size</span><span class="p">(</span><span class="mi">20000</span><span class="p">)</span></div>


<div class="viewcode-block" id="compute_binned_truth_sample_concordance"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.compute_binned_truth_sample_concordance">[docs]</a><span class="k">def</span> <span class="nf">compute_binned_truth_sample_concordance</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">binned_score_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">,</span>
        <span class="n">n_bins</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">100</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determines the concordance (TP, FP, FN) between a truth sample within the callset and the samples truth data</span>
<span class="sd">    grouped by bins computed using `compute_quantile_bin`.</span>

<span class="sd">    .. note::</span>

<span class="sd">        The input &#39;ht` should contain three row fields:</span>
<span class="sd">            - score: value to use for quantile binning</span>
<span class="sd">            - GT: a CallExpression containing the genotype of the evaluation data for the sample</span>
<span class="sd">            - truth_GT: a CallExpression containing the genotype of the truth sample</span>

<span class="sd">        The input `binned_score_ht` should contain:</span>
<span class="sd">             - score: value used to bin the full callset</span>
<span class="sd">             - bin: the full callset quantile bin</span>


<span class="sd">    The table is grouped by global/truth sample bin and variant type and contains TP, FP and FN.</span>

<span class="sd">    :param ht: Input HT</span>
<span class="sd">    :param binned_score_ht: Table with the an annotation for quantile bin for each variant</span>
<span class="sd">    :param n_bins: Number of bins to bin the data into</span>
<span class="sd">    :return: Binned truth sample concordance HT</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Annotate score and global bin</span>
    <span class="n">indexed_binned_score_ht</span> <span class="o">=</span> <span class="n">binned_score_ht</span><span class="p">[</span><span class="n">ht</span><span class="o">.</span><span class="n">key</span><span class="p">]</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">score</span><span class="o">=</span><span class="n">indexed_binned_score_ht</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
        <span class="n">global_bin</span><span class="o">=</span><span class="n">indexed_binned_score_ht</span><span class="o">.</span><span class="n">bin</span>
    <span class="p">)</span>

    <span class="c1"># Annotate the truth sample quantile bin</span>
    <span class="n">bin_ht</span> <span class="o">=</span> <span class="n">compute_quantile_bin</span><span class="p">(</span>
        <span class="n">ht</span><span class="p">,</span>
        <span class="n">score_expr</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">,</span>
        <span class="n">bin_expr</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;truth_sample_bin&#39;</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">expr</span><span class="o">.</span><span class="n">bool</span><span class="p">(</span><span class="kc">True</span><span class="p">)},</span>
        <span class="n">n_bins</span><span class="o">=</span><span class="n">n_bins</span>
    <span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">bin_ht</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">)</span>

    <span class="c1"># Explode the global and truth sample bins</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="nb">bin</span><span class="o">=</span><span class="p">[</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="s1">&#39;global_bin&#39;</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">global_bin</span><span class="p">]),</span>
        <span class="n">hl</span><span class="o">.</span><span class="n">tuple</span><span class="p">([</span><span class="s1">&#39;truth_sample_bin&#39;</span><span class="p">,</span> <span class="n">ht</span><span class="o">.</span><span class="n">truth_sample_bin</span><span class="p">])</span>
    <span class="p">])</span>

    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">bin</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">bin_id</span><span class="o">=</span><span class="n">ht</span><span class="o">.</span><span class="n">bin</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
        <span class="nb">bin</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">int</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">bin</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="c1"># Compute TP, FP and FN by bin_id, variant type and bin</span>
    <span class="k">return</span> <span class="n">ht</span><span class="o">.</span><span class="n">group_by</span><span class="p">(</span>
        <span class="s1">&#39;bin_id&#39;</span><span class="p">,</span>
        <span class="s1">&#39;snv&#39;</span><span class="p">,</span>
        <span class="s1">&#39;bin&#39;</span>
    <span class="p">)</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
        <span class="c1"># TP =&gt; allele is found in both data sets</span>
        <span class="n">tp</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">ht</span><span class="o">.</span><span class="n">truth_GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()),</span>
        <span class="c1"># FP =&gt; allele is found only in test data set</span>
        <span class="n">fp</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_else</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">truth_GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">(),</span> <span class="kc">True</span><span class="p">)),</span>
        <span class="c1"># FN =&gt; allele is found in truth data only</span>
        <span class="n">fn</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count_where</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">GT</span><span class="o">.</span><span class="n">is_hom_ref</span><span class="p">()</span> <span class="o">&amp;</span> <span class="n">hl</span><span class="o">.</span><span class="n">or_else</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">truth_GT</span><span class="o">.</span><span class="n">is_non_ref</span><span class="p">(),</span> <span class="kc">True</span><span class="p">)),</span>
        <span class="n">min_score</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">),</span>
        <span class="n">max_score</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">score</span><span class="p">),</span>
        <span class="n">n_alleles</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
    <span class="p">)</span><span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span></div>


<div class="viewcode-block" id="create_truth_sample_ht"><a class="viewcode-back" href="../../../api_reference/utils/variant_qc.html#gnomad.utils.variant_qc.create_truth_sample_ht">[docs]</a><span class="k">def</span> <span class="nf">create_truth_sample_ht</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">truth_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span>
        <span class="n">high_confidence_intervals_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Computes a table comparing a truth sample in callset vs the truth.</span>

<span class="sd">    :param mt: MT of truth sample from callset to be compared to truth</span>
<span class="sd">    :param truth_mt: MT of truth sample</span>
<span class="sd">    :param high_confidence_intervals_ht: High confidence interval HT</span>
<span class="sd">    :return: Table containing both the callset truth sample and the truth data</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">split_filter_and_flatten_ht</span><span class="p">(</span><span class="n">truth_mt</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">MatrixTable</span><span class="p">,</span> <span class="n">high_confidence_intervals_ht</span><span class="p">:</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">hl</span><span class="o">.</span><span class="n">Table</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Splits a truth sample MT and filter it to the given high confidence intervals.</span>
<span class="sd">        Then &quot;flatten&quot; it as a HT by annotating GT in a row field.</span>

<span class="sd">        :param truth_mt: Truth sample MT</span>
<span class="sd">        :param high_confidence_intervals_ht: High confidence intervals</span>
<span class="sd">        :return: Truth sample table with GT as a row annotation</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span><span class="p">(</span><span class="n">truth_mt</span><span class="o">.</span><span class="n">count_cols</span><span class="p">()</span> <span class="o">==</span> <span class="mi">1</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="s1">&#39;was_split&#39;</span> <span class="ow">in</span> <span class="n">truth_mt</span><span class="o">.</span><span class="n">row</span><span class="p">:</span>
            <span class="n">truth_mt</span> <span class="o">=</span> <span class="n">hl</span><span class="o">.</span><span class="n">split_multi_hts</span><span class="p">(</span><span class="n">truth_mt</span><span class="p">)</span>

        <span class="n">truth_mt</span> <span class="o">=</span> <span class="n">truth_mt</span><span class="o">.</span><span class="n">filter_rows</span><span class="p">(</span>
            <span class="n">hl</span><span class="o">.</span><span class="n">is_defined</span><span class="p">(</span><span class="n">high_confidence_intervals_ht</span><span class="p">[</span><span class="n">truth_mt</span><span class="o">.</span><span class="n">locus</span><span class="p">])</span>
        <span class="p">)</span>
        <span class="n">truth_mt</span> <span class="o">=</span> <span class="n">truth_mt</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="s1">&#39;GT&#39;</span><span class="p">:</span> <span class="s1">&#39;_GT&#39;</span><span class="p">})</span>
        <span class="k">return</span> <span class="n">truth_mt</span><span class="o">.</span><span class="n">annotate_rows</span><span class="p">(</span><span class="n">GT</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">agg</span><span class="o">.</span><span class="n">take</span><span class="p">(</span><span class="n">truth_mt</span><span class="o">.</span><span class="n">_GT</span><span class="p">,</span> <span class="mi">1</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">rows</span><span class="p">()</span>

    <span class="c1"># Load truth sample MT,</span>
    <span class="c1"># restrict it to high confidence intervals</span>
    <span class="c1"># and flatten it to a HT by annotating GT in a row annotation</span>
    <span class="n">truth_ht</span> <span class="o">=</span> <span class="n">split_filter_and_flatten_ht</span><span class="p">(</span>
        <span class="n">truth_mt</span><span class="p">,</span>
        <span class="n">high_confidence_intervals_ht</span>
    <span class="p">)</span>
    <span class="n">truth_ht</span> <span class="o">=</span> <span class="n">truth_ht</span><span class="o">.</span><span class="n">rename</span><span class="p">({</span><span class="n">f</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;truth_</span><span class="si">{f}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">truth_ht</span><span class="o">.</span><span class="n">row_value</span><span class="p">})</span>

    <span class="c1">#  Similarly load, filter and flatten callset truth sample MT</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">split_filter_and_flatten_ht</span><span class="p">(</span>
        <span class="n">mt</span><span class="p">,</span>
        <span class="n">high_confidence_intervals_ht</span>
    <span class="p">)</span>

    <span class="c1"># Outer join of truth and callset truth and annotate the score and global bin</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">truth_ht</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ht</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">)</span>
    <span class="n">ht</span> <span class="o">=</span> <span class="n">ht</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span>
        <span class="n">snv</span><span class="o">=</span><span class="n">hl</span><span class="o">.</span><span class="n">is_snp</span><span class="p">(</span><span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">ht</span><span class="o">.</span><span class="n">alleles</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">ht</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>

    </p>
  </div> 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>